---
title: "Gene-regulatory network of myocardiocyte after myocardial infarction"
output: html_document
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

During this tutorial, we will integrate the snRNA-seq and snATAC-seq data
generated from the human heart samples after myocardial infarction. The integrated
data will be used as input for inferring gene regulatory network.

We first download the required data. In this case, we need two Seurat objects 
with each one corresponding to snRNA-seq and snATAC-seq respectively. 
The snRNA-seq object includes gene expression data of all fibroblasts and 
the snATAC-seq includes all chromatin accessibility profiles. 
Additionally, we also need a gene activity matrix for data integration. 
This matrix was estimated from the snATAC-seq data by using 
the [ArchR](https://www.archrproject.com/) package. 
The script of cleaning the data and preparing these objects is found [here](https://costalab.ukaachen.de/open_data/scMEGA/Fibroblast/01_prepare_data.html).

Run the following commands to download the data:
```{bash, eval=FALSE}
mkdir -p Cardiomyocyte
wget -P Cardiomyocyte https://costalab.ukaachen.de/open_data/scMEGA/Cardiomyocyte/snRNA.rds
wget -P Cardiomyocyte https://costalab.ukaachen.de/open_data/scMEGA/Cardiomyocyte/snATAC.rds
wget -P Cardiomyocyte https://costalab.ukaachen.de/open_data/scMEGA/Cardiomyocyte/gene.activity.rds
```

Next, we load all necessary packages:
```{r}
suppressMessages(library(ArchR))
suppressMessages(library(Seurat))
suppressMessages(library(Signac))
suppressMessages(library(scMEGA))
suppressMessages(library(harmony))
suppressMessages(library(Nebulosa))
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
```

Let's load the data into memory and see how they look like

```{r}
obj.rna <- readRDS("./Cardiomyocyte/snRNA.rds")
obj.atac <- readRDS("./Cardiomyocyte/snATAC.rds")
gene.activity <- readRDS("./Cardiomyocyte/gene.activity.rds")

obj.rna
obj.atac
```

We can observe that there are 45,515 and 6,481 cells in our snRNA-seq and snATAC-seq datasets. We now visualize the data as colored by patients. Note that here we used the UMAP embedding generated from batch-corrected low-dimensional space so that no batch effects are observed from the 2D visualization.

```{r, fig.height = 5, fig.width = 12, fig.align = "center"}
p1 <- DimPlot(obj.rna, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snRNA-seq")
    
p2 <- DimPlot(obj.atac, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snATAC-seq")

p1 + p2
```

## Co-embedding
First, we need to project the data into a common low-dimensional space. This is done by using the CCA method from [Seurat](https://satijalab.org/seurat/). To this end, we have wrapped several functions from Seurat into a single function _CoembedData_.

```{r}
obj.coembed <- CoembedData(
  obj.rna,
  obj.atac, 
  gene.activity, 
  weight.reduction = "harmony", 
  verbose = FALSE
)
```

We next visualize the snRNA-seq and snATAC-seq in this shared UMAP space. The cells are colored by
patients or modalities.

```{r, fig.height = 5, fig.width = 12, fig.align = "center"}

p1 <- DimPlot(obj.coembed, group.by = "patient", shuffle = TRUE, label = TRUE)
p2 <- DimPlot(obj.coembed, group.by = "tech", shuffle = TRUE, label = TRUE)

p1 + p2
```

The batch effects between patients, regions and modalities are quite clear. 
So next we use [Harmony](https://www.nature.com/articles/s41592-019-0619-0) to 
perform batch correction and generate a new UMAP embedding.

```{r, fig.height = 10, fig.width = 12, fig.align = "center"}
obj.coembed <- RunHarmony(
  obj.coembed,
  group.by.vars = c("patient", "region", "tech"),
  reduction = "pca",
  max.iter.harmony = 30,
  dims.use = 1:30,
  project.dim = FALSE,
  plot_convergence = FALSE
)

obj.coembed <- RunUMAP(
  obj.coembed,
  dims = 1:30,
  reduction = 'harmony',
  reduction.name = "umap_harmony",
  reduction.ke = 'umapharmony_',
  verbose = FALSE,
  min.dist = 0.4
)
```

We can plot the data again

```{r, fig.height = 5, fig.width = 12, fig.align = "center"}
p1 <-
  DimPlot(obj.coembed, group.by = "patient", reduction = "umap_harmony")
p2 <-
  DimPlot(obj.coembed, group.by = "tech", reduction = "umap_harmony")

p1 + p2
```

From the new UMAP embedding, we can observe that after batch-correction, cells from different patients, regions, and modalities are well mixed.


## Sub-clustering

We next perform sub-clustering to identify different populations in our multi-omic fibroblast data. To further control the data quality, here we will use a two-round approach to remove low-quality cells. We first use a high-resolution to get a large amount of clusters.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
obj.coembed <- FindNeighbors(obj.coembed, reduction = "harmony", dims = 1:30)
obj.coembed <- FindClusters(obj.coembed, resolution = 1, verbose = FALSE)

cols <- ArchR::paletteDiscrete(obj.coembed@meta.data[, "RNA_snn_res.1"])
    
p <- DimPlot(obj.coembed, group.by = "RNA_snn_res.1", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE) +
    scale_color_manual(values = cols) +
    xlab("UMAP1") + ylab("UMAP2")
    
p
```

We can use the function _CellPropPlot_ to visualize the cell propotion across all patients.

```{r, fig.height = 6, fig.width = 10, fig.align = "center"}
cols <- ArchR::paletteDiscrete(obj.coembed@meta.data[, "patient_region_id"])

p1 <- CellPropPlot(obj.coembed, 
                   group.by = "patient_region_id", 
                   prop.in = "RNA_snn_res.1",
                  cols = cols)

p2 <- CellPropPlot(obj.coembed, 
                   group.by = "tech", 
                   prop.in = "RNA_snn_res.1")

p1
p2
```


Next, we identify the markers for each cluster and visualize the top 3.

```{r, fig.height = 5, fig.width = 15, fig.align = "center", eval=TRUE}
all.markers <- FindAllMarkers(obj.coembed, 
                              only.pos = TRUE, 
                              min.pct = 0.5, logfc.threshold = 0.5)

df <- all.markers %>%
    group_by(cluster) %>%
    slice_max(n = 3, order_by = avg_log2FC)

p <- DotPlot(obj.coembed, features = unique(df$gene)) + RotatedAxis()

print(p)

saveRDS(obj.coembed, "./Cardiomyocyte/coembed.rds")
```

We remove the clusters that are either patient-specific or modality-specific.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
Idents(obj.coembed) <- "RNA_snn_res.1"
coembed.sub <- subset(obj.coembed, idents = c(1, 2, 21, 22, 23, 24, 26), invert = TRUE)
coembed.sub

cols.clusters <- ArchR::paletteDiscrete(coembed.sub@meta.data[, "RNA_snn_res.1"])

options(repr.plot.height = 6, repr.plot.width = 6)

p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.1", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE, cols = cols.clusters) +
    xlab("UMAP1") + ylab("UMAP2")

p
```

We then re-do the UMAP embedding and clustering with a lower resolution to reduce complexity.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
coembed.sub <- RunUMAP(coembed.sub, 
               dims = 1:30, 
               reduction = 'harmony',
               reduction.name = "umap_harmony",
               reduction.key = 'umap_harmony_',
              verbose = FALSE,
                   min.dist = 0.4)

## re-clustering
coembed.sub <- FindNeighbors(coembed.sub, reduction = "harmony", dims = 1:30)
coembed.sub <- FindClusters(coembed.sub, resolution = 0.1, verbose = FALSE)

cols <- ArchR::paletteDiscrete(coembed.sub@meta.data[, "RNA_snn_res.0.1"])
    
p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.1", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE, cols = cols) +
    xlab("UMAP1") + ylab("UMAP2")
    
p              
```

Marker genes are identified based on new clustering results and we can plot the
top 10 markers.

```{r, fig.height = 5, fig.width = 15, fig.align = "center"}
all.markers <- FindAllMarkers(coembed.sub, 
                              only.pos = TRUE, 
                              min.pct = 0.25, logfc.threshold = 0.5)

df <- all.markers %>%
    group_by(cluster) %>%
    slice_max(n = 10, order_by = avg_log2FC)

p <- DotPlot(coembed.sub, features = unique(df$gene)) + RotatedAxis()

print(p)

saveRDS(coembed.sub, "./Cardiomyocyte/coembed.cluster.rds")
```

We can plot the snRNA-seq and snATAC-seq data separately

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.1", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE, split.by = "tech", 
             cols = cols) +
    xlab("UMAP1") + ylab("UMAP2")

p
```

Visualize the cell proportion of each sub-population across all patients

```{r, fig.height = 6, fig.width = 10, fig.align = "center"}
p <- CellPropPlot(coembed.sub, 
                   group.by = "RNA_snn_res.0.1", 
                   prop.in = "patient_region_id", 
                   cols = cols)

p
```

Since we have annotated all patients into three major groups, i.e., myogenic, ischmeic, and fibrotic.
we can also perform statistical test to check if any sub-population are enriched in any of the above group. This can be done by the function _CompareCellProp_.

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
coembed.sub$patient_group <- factor(coembed.sub$patient_group, 
                                    levels = c("myogenic", "ischemic", "fibrotic"))

p <- CompareCellProp(object = coembed.sub, 
                     group.by = "RNA_snn_res.0.1", 
                     prop.in = "patient_region_id", 
                      sample.annotation = "patient_group",
                    comparisons = list(c("myogenic", "ischemic"),
                                       c("ischemic", "fibrotic"),
                                       c("myogenic", "fibrotic")))

p
```

Save data
```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
saveRDS(coembed.sub, "./Cardiomyocyte/coembed.sub.rds")
```

```{r}
# Check session information
sessionInfo()
```

