---
title: "Gene-regulatory network of myofibroblast differentiation in myocardial infarction"
output: html_document
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

During this tutorial, we will be analyzing the snRNA-seq and snATAC-seq generated from human hearts after myocardial infarction to understand the gene regulatory dynamics for myofibroblast differentiation. More specifically, we will:  

* Pair the cells between gene expression and chromatin accessibility data
* Identify the trajectory for myofibroblast differentiation
* Select the relevant TFs and genes for this process
* Infer and visualize the gene regulatory network

We first download all the required data. In this case, we need two Seurat objects with each one corresponding to snRNA-seq and snATAC-seq respectively. The snRNA-seq object includes gene expression data of all fibroblasts and snRNA-seq includes all chromatin accessibility data. Additionally, we also need a gene activity matrix for data integration. This matrix was estimated from the snATAC-seq data by using [ArchR](https://www.archrproject.com/) package. The script of cleaning the data and preparing these objects is found [here](https://www.dropbox.com/s/9i1lni0iia71ape/01_prepare_data.html?dl=0).

Run the following commands to download the data:
```{bash, eval=FALSE}
wget https://www.dropbox.com/s/6v98iiq8dtlcjho/snRNA.rds
wget https://www.dropbox.com/s/h30ej52burtsgdy/snATAC.rds
wget https://www.dropbox.com/s/7uc2ad0jcrzks20/gene.activity.rds
```

Next, we load all necessary packages:
```{r}
suppressMessages(library(Seurat))
suppressMessages(library(Signac))
suppressMessages(library(scMEGA))
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(JASPAR2020))
suppressMessages(library(TFBSTools))
suppressMessages(library(ggplot2))
```

Let's load the data into memory and see how they look like
```{r}
obj.rna <- readRDS("./snRNA.rds")
obj.atac <- readRDS("./snATAC.rds")

obj.rna
obj.atac
```

We can observe that there are 45,515 and 6,481 cells in our snRNA-seq and snATAC-seq datasets. We now visualize the data as colored by patients. Note that here we used the UMAP embedding generated from batch-corrected low-dimensional space so that no batch effects are observed from the 2D visualization.
```{r, fig.height = 5, fig.width = 12, fig.align = "center"}
p1 <- DimPlot(obj.rna, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snRNA-seq")
    
p2 <- DimPlot(obj.atac, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snATAC-seq")

p1 + p2
```

## Cell pairing
In this step, we match the cells between these two modalities. In other words, for each cell in, for example, snATAC-seq, we will find a cell from snRNA-seq data so that these two cells have the similar profiles.

### Co-embedding
First, we need to project the data into a common low-dimensional space. This is done by using the CCA method from [Seurat](https://satijalab.org/seurat/). More specifically, we wrapped several functions from Seurat into a single function _CoembedData_.

## Trajectory identification
## TFs and genes selection
## Gene regulatory network inference and visualization


```{r}
# Check session information
sessionInfo()
```

