---
title: "Gene-regulatory network of myofibroblast differentiation in myocardial infarction"
output: html_document
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

During this tutorial, we will be analyzing the snRNA-seq and snATAC-seq generated from human hearts after myocardial infarction to understand the gene regulatory dynamics for myofibroblast differentiation. More specifically, we will:  

* Pair the cells between gene expression and chromatin accessibility data
* Identify the trajectory for myofibroblast differentiation
* Select the relevant TFs and genes for this process
* Infer and visualize the gene regulatory network

We first download all the required data. In this case, we need two Seurat objects with each one corresponding to snRNA-seq and snATAC-seq respectively. The snRNA-seq object includes gene expression data of all fibroblasts and snRNA-seq includes all chromatin accessibility data. Additionally, we also need a gene activity matrix for data integration. This matrix was estimated from the snATAC-seq data by using [ArchR](https://www.archrproject.com/) package. The script of cleaning the data and preparing these objects is found [here](https://www.dropbox.com/s/9i1lni0iia71ape/01_prepare_data.html?dl=0).

Run the following commands to download the data:
```{bash, eval=FALSE}
wget https://www.dropbox.com/s/6v98iiq8dtlcjho/snRNA.rds
wget https://www.dropbox.com/s/h30ej52burtsgdy/snATAC.rds
wget https://www.dropbox.com/s/7uc2ad0jcrzks20/gene.activity.rds
```

Next, we load all necessary packages:
```{r}
suppressMessages(library(Seurat))
suppressMessages(library(Signac))
suppressMessages(library(scMEGA))
suppressMessages(library(harmony))
suppressMessages(library(Nebulosa))
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(JASPAR2020))
suppressMessages(library(TFBSTools))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
```

Let's load the data into memory and see how they look like
```{r}
obj.rna <- readRDS("./snRNA.rds")
obj.atac <- readRDS("./snATAC.rds")
gene.activity <- readRDS("./gene.activity.rds")

obj.rna
obj.atac
```

We can observe that there are 45,515 and 6,481 cells in our snRNA-seq and snATAC-seq datasets. We now visualize the data as colored by patients. Note that here we used the UMAP embedding generated from batch-corrected low-dimensional space so that no batch effects are observed from the 2D visualization.
```{r, fig.height = 5, fig.width = 12, fig.align = "center"}
p1 <- DimPlot(obj.rna, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snRNA-seq")
    
p2 <- DimPlot(obj.atac, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snATAC-seq")

p1 + p2
```

## Cell pairing
In this step, we match the cells between these two modalities. In other words, for each cell in, for example, snATAC-seq, we will find a cell from snRNA-seq data so that these two cells have the similar profiles.

### Co-embedding
First, we need to project the data into a common low-dimensional space. This is done by using the CCA method from [Seurat](https://satijalab.org/seurat/). More specifically, we wrapped several functions from Seurat into a single function _CoembedData_.

```{r, fig.height = 5, fig.width = 12, fig.align = "center"}
obj.coembed <- CoembedData(obj.rna, obj.atac, gene.activity, 
                           weight.reduction = "harmony", 
                           verbose = TRUE)
```

We next visualize the snRNA-seq and snATAC-seq in this shared space. The cells are coloded by
patient, region, modality, or patient group.
```{r, fig.height = 10, fig.width = 12, fig.align = "center"}

p1 <- DimPlot(obj.coembed, group.by = "patient", shuffle = TRUE, label = TRUE)
p2 <- DimPlot(obj.coembed, group.by = "region", shuffle = TRUE, label = TRUE)
p3 <- DimPlot(obj.coembed, group.by = "tech", shuffle = TRUE, label = TRUE)
p4 <- DimPlot(obj.coembed, group.by = "patient_group", shuffle = TRUE, label = TRUE)

(p1 + p2) / (p3 + p4)
```

The batch effects between patients, regions and modalities are quite clear. 
So next we use [Harmony](https://www.nature.com/articles/s41592-019-0619-0) to 
perform batch correction and generate a new UMAP embedding.

```{r, fig.height = 10, fig.width = 12, fig.align = "center"}
obj.coembed <- RunHarmony(obj.coembed, 
                      group.by.vars = c("patient", "region", "tech"),
                     reduction = "pca", 
                      max.iter.harmony = 30, 
                      dims.use = 1:30,
                     project.dim = FALSE,
                     plot_convergence = FALSE)

obj.coembed <- RunUMAP(obj.coembed, 
               dims = 1:30, 
               reduction = 'harmony',
               reduction.name = "umap_harmony",
               reduction.ke = 'umapharmony_',
              verbose = FALSE,
                   min.dist = 0.4)
```

We can plot the data again
```{r, fig.height = 10, fig.width = 12, fig.align = "center"}

p1 <- DimPlot(obj.coembed, group.by = "patient", reduction = "umap_harmony")
p2 <- DimPlot(obj.coembed, group.by = "region", reduction = "umap_harmony")
p3 <- DimPlot(obj.coembed, group.by = "tech", reduction = "umap_harmony")
p4 <- DimPlot(obj.coembed, group.by = "patient_group", reduction = "umap_harmony")

(p1 + p2) / (p3 + p4)
```
The new UMAP embedding shows that after batch-correction, cells from different patients, regions, and modalities are well mixed.


Based on our previous works of myofibroblast differentiation in [human](https://www.nature.com/articles/s41586-020-2941-1) and [mouse](https://www.nature.com/articles/s41467-021-26530-2) kidney, we already known some relevant genes for this biological process. For example, SCARA5 is a marker for myofibroblast progenitor, and COL1A1, POSTN, and FN1 are highly expressed in myofibroblast. Therefore we can visualize the expression of these genes to check if we can also identify similar process in human heart. Note that to make the visualization clear, here we used the package [Nebulosa](https://github.com/powellgenomicslab/Nebulosa) to plot the data.

```{r, fig.height = 10, fig.width = 12, fig.align = "center"}

p1 <- plot_density(obj.coembed, features="SCARA5", reduction="umap_harmony", pal = "magma")
p2 <- plot_density(obj.coembed, features="COL1A1", reduction="umap_harmony", pal = "magma")
p3 <- plot_density(obj.coembed, features="POSTN", reduction="umap_harmony", pal = "magma")
p4 <- plot_density(obj.coembed, features="FN1", reduction="umap_harmony", pal = "magma")


(p1 + p2) / (p3 + p4)
```
From the visulization, we can observe that some cells highly express SCARA5 which could be the progentiors of myofibroblasts. On the other hand, some cells highly express COL1A1, POSTN, and FN1 and they could be terminally differentiated myofibroblasts.

### Sub-clustering and cleaning
We next perform sub-clustering to identify different populations in our fibroblast data. To further control the data quality, here we will use a two-round approach to remove low-quality cells. We first use a high-resolution to get a large amount of clusters.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
obj.coembed <- FindNeighbors(obj.coembed, reduction = "harmony", dims = 1:30)
obj.coembed <- FindClusters(obj.coembed, resolution = 0.9, verbose = FALSE)

cols <- ArchR::paletteDiscrete(obj.coembed@meta.data[, "RNA_snn_res.0.9"])
    
p <- DimPlot(obj.coembed, group.by = "RNA_snn_res.0.9", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE) +
    scale_color_manual(values = cols) +
    xlab("UMAP1") + ylab("UMAP2")
    
p
```

We can use the function _CellPropPlot_ to visualize the cell propotion across all patients.

```{r, fig.height = 6, fig.width = 10, fig.align = "center"}
p <- CellPropPlot(obj.coembed, 
                   group.by = "RNA_snn_res.0.9", 
                   prop.in = "patient_region_id", 
                   cols = cols)

p
```

Next, we identify the markers for each cluster and visualize the top 5.
```{r, fig.height = 5, fig.width = 15, fig.align = "center"}
all.markers <- FindAllMarkers(obj.coembed, 
                              only.pos = TRUE, 
                              min.pct = 0.25, logfc.threshold = 0.5)

df <- all.markers %>%
    group_by(cluster) %>%
    slice_max(n = 3, order_by = avg_log2FC)

p <- DotPlot(obj.coembed, features = unique(df$gene)) + RotatedAxis()

print(p)
```


## Trajectory identification
## TFs and genes selection
## Gene regulatory network inference and visualization


```{r}
# Check session information
sessionInfo()
```

