---
title: "Gene-regulatory network of myofibroblast differentiation in myocardial infarction"
output: html_document
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

During this tutorial, we will be analyzing the snRNA-seq and snATAC-seq generated from human hearts after myocardial infarction to understand the gene regulatory dynamics for myofibroblast differentiation. More specifically, we will:  

* Integrate the snRNA-seq and snATAC data
* Identify the trajectory for myofibroblast differentiation
* Select the relevant TFs and genes for this process
* Infer and visualize the gene regulatory network

We first download all the required data. In this case, we need two Seurat objects with each one corresponding to snRNA-seq and snATAC-seq respectively. The snRNA-seq object includes gene expression data of all fibroblasts and snRNA-seq includes all chromatin accessibility data. Additionally, we also need a gene activity matrix for data integration. This matrix was estimated from the snATAC-seq data by using [ArchR](https://www.archrproject.com/) package. The script of cleaning the data and preparing these objects is found [here](https://www.dropbox.com/s/9i1lni0iia71ape/01_prepare_data.html?dl=0).

Run the following commands to download the data:
```{bash, eval=FALSE}
wget https://www.dropbox.com/s/6v98iiq8dtlcjho/snRNA.rds
wget https://www.dropbox.com/s/h30ej52burtsgdy/snATAC.rds
wget https://www.dropbox.com/s/7uc2ad0jcrzks20/gene.activity.rds
```

Next, we load all necessary packages:
```{r}
suppressMessages(library(ArchR))
suppressMessages(library(Seurat))
suppressMessages(library(Signac))
suppressMessages(library(scMEGA))
suppressMessages(library(harmony))
suppressMessages(library(Nebulosa))
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(JASPAR2020))
suppressMessages(library(TFBSTools))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
```

Let's load the data into memory and see how they look like
```{r}
obj.rna <- readRDS("./snRNA.rds")
obj.atac <- readRDS("./snATAC.rds")
gene.activity <- readRDS("./gene.activity.rds")

obj.rna
obj.atac
```

We can observe that there are 45,515 and 6,481 cells in our snRNA-seq and snATAC-seq datasets. We now visualize the data as colored by patients. Note that here we used the UMAP embedding generated from batch-corrected low-dimensional space so that no batch effects are observed from the 2D visualization.
```{r, fig.height = 5, fig.width = 12, fig.align = "center"}
p1 <- DimPlot(obj.rna, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snRNA-seq")
    
p2 <- DimPlot(obj.atac, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snATAC-seq")

p1 + p2
```

## Data integration
In this step, we integrate the snRNA-seq and snATAC-seq so that we can define the sample sub-popualtions between these two modalities.

### Co-embedding
First, we need to project the data into a common low-dimensional space. This is done by using the CCA method from [Seurat](https://satijalab.org/seurat/). To this end, we have wrapped several functions from Seurat into a single function _CoembedData_.

```{r}
obj.coembed <- CoembedData(obj.rna, obj.atac, gene.activity, 
                           weight.reduction = "harmony", 
                           verbose = FALSE)
```

We next visualize the snRNA-seq and snATAC-seq in this shared UMAP space. The cells are coloded by
patient, region, modality, or patient group.
```{r, fig.height = 10, fig.width = 12, fig.align = "center"}

p1 <- DimPlot(obj.coembed, group.by = "patient", shuffle = TRUE, label = TRUE)
p2 <- DimPlot(obj.coembed, group.by = "region", shuffle = TRUE, label = TRUE)
p3 <- DimPlot(obj.coembed, group.by = "tech", shuffle = TRUE, label = TRUE)
p4 <- DimPlot(obj.coembed, group.by = "patient_group", shuffle = TRUE, label = TRUE)

(p1 + p2) / (p3 + p4)
```

The batch effects between patients, regions and modalities are quite clear. 
So next we use [Harmony](https://www.nature.com/articles/s41592-019-0619-0) to 
perform batch correction and generate a new UMAP embedding.

```{r, fig.height = 10, fig.width = 12, fig.align = "center"}
obj.coembed <- RunHarmony(obj.coembed, 
                      group.by.vars = c("patient", "region", "tech"),
                     reduction = "pca", 
                      max.iter.harmony = 30, 
                      dims.use = 1:30,
                     project.dim = FALSE,
                     plot_convergence = FALSE)

obj.coembed <- RunUMAP(obj.coembed, 
               dims = 1:30, 
               reduction = 'harmony',
               reduction.name = "umap_harmony",
               reduction.ke = 'umapharmony_',
              verbose = FALSE,
                   min.dist = 0.4)
```

We can plot the data again

```{r, fig.height = 10, fig.width = 12, fig.align = "center"}

p1 <- DimPlot(obj.coembed, group.by = "patient", reduction = "umap_harmony")
p2 <- DimPlot(obj.coembed, group.by = "region", reduction = "umap_harmony")
p3 <- DimPlot(obj.coembed, group.by = "tech", reduction = "umap_harmony")
p4 <- DimPlot(obj.coembed, group.by = "patient_group", reduction = "umap_harmony")

(p1 + p2) / (p3 + p4)
```
From the new UMAP embedding, we can observe that after batch-correction, cells from different patients, regions, and modalities are well mixed.


Based on our previous works of myofibroblast differentiation in [human](https://www.nature.com/articles/s41586-020-2941-1) and [mouse](https://www.nature.com/articles/s41467-021-26530-2) kidney, we already known some relevant genes for this biological process. For example, SCARA5 is a marker for myofibroblast progenitor, and COL1A1, POSTN, and FN1 are highly expressed in myofibroblast. Therefore we can visualize the expression of these genes to check if we can also identify similar process in human heart. Note that to make the visualization clear, here we used the package [Nebulosa](https://github.com/powellgenomicslab/Nebulosa) to plot the data.

```{r, fig.height = 10, fig.width = 12, fig.align = "center"}

p1 <- plot_density(obj.coembed, features="SCARA5", reduction="umap_harmony", pal = "magma")
p2 <- plot_density(obj.coembed, features="COL1A1", reduction="umap_harmony", pal = "magma")
p3 <- plot_density(obj.coembed, features="POSTN", reduction="umap_harmony", pal = "magma")
p4 <- plot_density(obj.coembed, features="FN1", reduction="umap_harmony", pal = "magma")


(p1 + p2) / (p3 + p4)
```
From the visulization, we can observe that some cells highly express SCARA5 which could be the progentiors of myofibroblasts. On the other hand, some cells highly express COL1A1, POSTN, and FN1 and they could be terminally differentiated myofibroblasts.

### Sub-clustering

We next perform sub-clustering to identify different populations in our multi-omic fibroblast data. To further control the data quality, here we will use a two-round approach to remove low-quality cells. We first use a high-resolution to get a large amount of clusters.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
obj.coembed <- FindNeighbors(obj.coembed, reduction = "harmony", dims = 1:30)
obj.coembed <- FindClusters(obj.coembed, resolution = 0.9, verbose = FALSE)

cols <- ArchR::paletteDiscrete(obj.coembed@meta.data[, "RNA_snn_res.0.9"])
    
p <- DimPlot(obj.coembed, group.by = "RNA_snn_res.0.9", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE) +
    scale_color_manual(values = cols) +
    xlab("UMAP1") + ylab("UMAP2")
    
p
```

We can use the function _CellPropPlot_ to visualize the cell propotion across all patients.

```{r, fig.height = 6, fig.width = 10, fig.align = "center"}
p <- CellPropPlot(obj.coembed, 
                   group.by = "RNA_snn_res.0.9", 
                   prop.in = "patient_region_id", 
                   cols = cols)

p
```

Next, we identify the markers for each cluster and visualize the top 3.

```{r, fig.height = 5, fig.width = 15, fig.align = "center", eval=TRUE}
all.markers <- FindAllMarkers(obj.coembed, 
                              only.pos = TRUE, 
                              min.pct = 0.25, logfc.threshold = 0.5)

df <- all.markers %>%
    group_by(cluster) %>%
    slice_max(n = 3, order_by = avg_log2FC)

p <- DotPlot(obj.coembed, features = unique(df$gene)) + RotatedAxis()

print(p)
```
The above dot plot demonstrates the top 3 markers per cluster and we can easily classify cluster
3 and 7 as myofibroblasts. In addition, two clusters (i.e., cluster 2 and 5) are marked by RYR2, a well known marker gene for cardiomyocyte, likely due to the background noise. It is worth pointing out that in another study [Cells of the adult human heart](https://www.nature.com/articles/s41586-020-2797-4) a similar fibroblast sub-population (marked by another cardiomyocyte-specific gene TNNT2) was also identified. Based on these, we decide to remove these two clusters.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
Idents(obj.coembed) <- "RNA_snn_res.0.9"
coembed.sub <- subset(obj.coembed, idents = c(2, 5), invert = TRUE)
coembed.sub

cols.clusters <- ArchR::paletteDiscrete(coembed.sub@meta.data[, "RNA_snn_res.0.9"])

p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.9", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE, cols = cols) +
    xlab("UMAP1") + ylab("UMAP2")

p
```

We then re-do the UMAP embedding and clustering with a lower resolution to reduce complexity.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
coembed.sub <- RunUMAP(coembed.sub, 
                       dims = 1:30, 
                       reduction = 'harmony',
                       reduction.name = "umap_harmony",
                       reduction.key = 'umap_harmony_',
                       verbose = FALSE,
                       min.dist = 0.4)

## re-clustering
coembed.sub <- FindNeighbors(coembed.sub, reduction = "harmony", dims = 1:30)
coembed.sub <- FindClusters(coembed.sub, resolution = 0.2, verbose = FALSE)

cols <- ArchR::paletteDiscrete(coembed.sub@meta.data[, "RNA_snn_res.0.2"])
    
p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.2", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE) +
    scale_color_manual(values = cols) +
    xlab("UMAP1") + ylab("UMAP2")
    
p                
```

Marker genes are identified based on new clustering results and we can plot the
top 10 markers.

```{r, fig.height = 5, fig.width = 15, fig.align = "center"}
all.markers <- FindAllMarkers(coembed.sub, 
                              only.pos = TRUE, 
                              min.pct = 0.25, logfc.threshold = 0.5)

df <- all.markers %>%
    group_by(cluster) %>%
    slice_max(n = 10, order_by = avg_log2FC)

p <- DotPlot(coembed.sub, features = unique(df$gene)) + RotatedAxis()

print(p)
```
Now the clustering results are clearer, and cluster 1 is myofibroblast, and cluste 2
is SCARA5+ fibroblast.


We can plot the snRNA-seq and snATAC-seq separately

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.2", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE, split.by = "tech", 
             cols = cols) +
    xlab("UMAP1") + ylab("UMAP2")

p
```

Visualize the cell proportion of each sub-population across all patients
```{r, fig.height = 6, fig.width = 10, fig.align = "center"}
p <- CellPropPlot(coembed.sub, 
                   group.by = "RNA_snn_res.0.2", 
                   prop.in = "patient_region_id", 
                   cols = cols)

p
```

Since we have annotated all patients into three major groups, i.e., myogenic, ischmeic, and fibrotic.
we can also perform statistical test to check if any sub-popualtion are enriched in any of the above group. This can be done by the function _CompareCellProp_.

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
coembed.sub$patient_group <- factor(coembed.sub$patient_group, 
                                    levels = c("myogenic", "ischemic", "fibrotic"))

p <- CompareCellProp(object = coembed.sub, 
                     group.by = "RNA_snn_res.0.2", 
                     prop.in = "patient_region_id", 
                      sample.annotation = "patient_group",
                    comparisons = list(c("myogenic", "ischemic"),
                                       c("ischemic", "fibrotic"),
                                       c("myogenic", "fibrotic")))

p
```

This analysis reveals significant changes of cell proportion between different conditions for cluster 1, 2, and 3. We therefore only use these three clusters in our trajectory analysis.

```{r}
Idents(coembed.sub) <- "RNA_snn_res.0.2"
coembed.sub <- subset(coembed.sub, idents = c(1, 2, 3))
```

We can generate a new UMAP

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
cols <- ArchR::paletteDiscrete(unique(coembed.sub@meta.data[, "RNA_snn_res.0.2"]))

coembed.sub <- RunUMAP(coembed.sub, 
               dims = 1:30, 
               reduction = 'harmony',
               reduction.name = "umap_harmony",
               reduction.ke = 'umap_harmony_',
              verbose = FALSE)

p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.2", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE, cols = cols) +
    xlab("UMAP1") + ylab("UMAP2")

p
```

## Trajectory identification

### Dimensionality reduction
To infer trajectory, we will perform dimension reduction using diffusion map via the function
_RunDiffusionMap_. This is based on the R package [destiny](https://bioconductor.org/packages/release/bioc/html/destiny.html).

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
coembed.sub <- RunDiffusionMap(coembed.sub, reduction = "harmony")

p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.2", label = TRUE,
             reduction = "dm", shuffle = TRUE, cols = cols) +
    xlab("DC 1") + ylab("DC 2")

p
```
We can also plot snATAC-seq and snRNA-seq individually

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
DimPlot(coembed.sub, reduction = "dm", 
        group.by = "RNA_snn_res.0.2", split.by = "tech", cols = cols)
```

### Cell pairing
Next, we match the cells between these two modalities. In other words, for each cell in, for example, snATAC-seq, we will find a cell from snRNA-seq data so that these two cells have the similar profiles. This is only necessary when each modalitiy was performed indepedently. If snRNA-seq and snATAC-seq data was generated by multi-modal protocol, e.g., 10X multiome or SHARE-seq, this step can be skipped.

We here use the method proposed by [Kartha, Vinay K., et al.](https://www.biorxiv.org/content/10.1101/2021.07.28.453784v1.abstract) to match the cells.

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
df.pair <- PairCells(object = coembed.sub, reduction = "harmony",
                    pair.by = "tech", ident1 = "ATAC", ident2 = "RNA")
```

We can visualize the paired cells

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
sel_cells <- c(df.pair$ATAC, df.pair$RNA)
coembed.sub2 <- coembed.sub[, sel_cells]

options(repr.plot.height = 5, repr.plot.width = 10)
DimPlot(coembed.sub2, reduction = "dm", 
        group.by = "RNA_snn_res.0.2", split.by = "tech", cols = cols)
```
We next create a new Seurat object for there paired cells as if they are generated by single-cell
multimodal protocol.

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
obj.pair <- CreatePairedObject(df.pair = df.pair, 
                               object = coembed.sub2,
                               use.assay1 = "RNA", 
                               use.assay2 = "ATAC")

obj.pair
```
Finally, we infer a pseudo-time trajectory from SCARA5+ fibroblasts to myofibroblast using the approach from [ArchR](https://www.archrproject.com/). Here we modified the function to allow to take a Seurat object as input

```{r, fig.height = 5, fig.width = 5, fig.align = "center"}
obj.pair <- AddTrajectory(object = obj.pair, 
                          trajectory = c(2, 1),
                          group.by = "RNA_snn_res.0.2", 
                          reduction = "dm",
                          dims = 1:3, 
                          use.all = FALSE)
                          
# we only plot the cells that are in this trajectory
obj <- obj.pair[, !is.na(obj.pair$Trajectory)]

p <- TrajectoryPlot(object = obj, 
                    reduction = "dm",
                    continuousSet = "blueYellow",
                    size = 1,
                   addArrow = FALSE)

p
```

## TF and gene selection
We next select candidate TFs and genes for building a meaningful gene regulatory network.

### Select TFs
To identify potential regulator (i.e., TFs), we first estimate an acitivty score for each TF in each cell. This is done by first performing motif matching and then computing deviation scores using [chromVAR](https://greenleaflab.github.io/chromVAR/index.html).

```{r}
# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
obj <- AddMotifs(
  object = obj,
  genome = BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm,
    assay = "ATAC"
)

obj <- RunChromVAR(
  object = obj,
  genome = BSgenome.Hsapiens.UCSC.hg38,
    assay = "ATAC"
)
```

We can visualize the TF activity dynamic along the trajectory

```{r, fig.height = 6, fig.width = 6, fig.align = "center"}
trajMM <- GetTrajectory(obj, 
                        assay = "chromvar", 
                        slot = "data",
                       smoothWindow = 7,
                       log2Norm = FALSE)

rownames(trajMM) <- obj@assays$ATAC@motifs@motif.names

ht <- TrajectoryHeatmap(trajMM, 
                        varCutOff = 0,
                        pal = paletteContinuous(set = "solarExtra"),
                        limits = c(-2, 2),
                       name = "TF activity")

ht
```

Apparently many TFs are showing dynamic TF activity along the trajectory. To reduce the number of candidates, we will use snRNA-seq to pick up the TFs that also show dynamic expression.

```{r}
trajRNA <- GetTrajectory(obj, 
                        assay = "RNA", 
                        slot = "data",
                       smoothWindow = 7,
                       log2Norm = TRUE)
                       
df_cor <- GetCorrelation(trajMM, trajRNA)

# we only select TFs that show significant correlation
df_cor <- df_cor[df_cor$adj_p < 0.01 & df_cor$correlation > 0.3, ]
```

We can try to assign a pseudo-time label for each TF.
```{r}
matMM <- TrajectoryHeatmap(trajMM, 
                        varCutOff = 0,
                        pal = paletteContinuous(set = "solarExtra"),
                        limits = c(-2, 2),
                       name = "TF activity",
                       returnMatrix = TRUE)

df_tf_time_point <- data.frame(tfs = rownames(matMM),
                               time_point = seq(1, 100, length.out = nrow(matMM)))
rownames(df_tf_time_point) <- df_tf_time_point$tfs

df_tf_time_point <- df_tf_time_point[df_cor$tfs, ]
df_cor$time_point <- df_tf_time_point$time_point
df_cor <- df_cor[order(df_cor$time_point), ]

head(df_cor)
```

We can visualize the TF activity and expression dynamics along the trajectory by generating a joint heatmap plot.
```{r, fig.height = 8, fig.width = 8, fig.align = "center"}
trajMM <- trajMM[df_cor$tfs, ]
trajRNA <- trajRNA[df_cor$tfs, ]

ht <- CorrelationHeatmap(trajectory1 = trajMM, 
                         trajectory2 = trajRNA,
                         name1 = "TF activity",
                         name2 = "Gene expression")

ht
```

### Select genes

We will select relevant genes by first linking genes to peaks based on the corrleation between gene expression from the snRNA-seq data and peak accessibility from the snATAC-seq data along the inferred trajectory. This means that we only consider a gene to be a potential target if it can be assocaited to at least one peak.

Let's first get the gene expression and chromatin accessibility data along our trajectory
```{r}
trajRNA <- GetTrajectory(obj, 
                        assay = "RNA", 
                        slot = "data",
                       smoothWindow = 7,
                       log2Norm = TRUE)

trajATAC <- GetTrajectory(obj, 
                        assay = "ATAC", 
                        slot = "data",
                       smoothWindow = 7,
                       log2Norm = TRUE)

# note here we only use the top 10% most variable genes
groupMatRNA <- TrajectoryHeatmap(trajRNA,
                        varCutOff = 0.9,
                        pal = paletteContinuous(set = "horizonExtra"),
                        limits = c(-2, 2),
                                 returnMatrix = TRUE)

groupMatATAC <- TrajectoryHeatmap(trajATAC, 
                        varCutOff = 0,
                                  maxFeatures = nrow(trajATAC),
                            pal = paletteContinuous(set = "solarExtra"),
                            limits = c(-2, 2),
                       name = "Chromatin accessibility",
                                 returnMatrix = TRUE)
```

The peak-to-gene link can be identified by using the function _PeakToGene_
```{r}
df.p2g <- PeakToGene(peak.mat = groupMatATAC,
                     gene.mat = groupMatRNA, 
                     genome = "hg38")
```

We here filter the links based on the following criteria:

* The peak must be a distal regulatory element which likely play an enhancter role
* The correlation between gene expression and peak accessibility muse be positive and significant

```{r}
df.p2g <- df.p2g %>%
    subset(distance > 2000) %>%
    subset(Correlation > 0 & FDR < 1e-04)
```

After filtering, we can also generate a joint heatmap to visualize the peak accessibility and associated gene expression dynamics along the trajectory
```{r, fig.height = 8, fig.width = 12, fig.align = "center"}
trajATAC <- trajATAC[df.p2g$peak, ]
trajRNA <- trajRNA[df.p2g$gene, ]

ht <- CorrelationHeatmap(trajectory1 = trajATAC, 
                         trajectory2 = trajRNA,
                        name1 = "Chromatin accessibility",
                        name2 = "Gene expression",
                        labelRows1 = FALSE,
                        labelRows2 = FALSE)

ht
```


## Gene regulatory network inference and visualization


```{r}
# Check session information
sessionInfo()
```

