---
title: "Gene-regulatory network of myofibroblast differentiation in myocardial infarction"
output: html_document
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this tutorial, we will use the integrated snRNA-seq and snATAC-seq data to infer a gene regulatory network to characterize myofibroblast differentiation. More details about
data integration can be found [here](https://costalab.github.io/scMEGA/articles/myofibroblast-preprocessing.html)

More specifically, we will:  

* Identify the trajectory for myofibroblast differentiation
* Select the relevant TFs and genes for this process
* Infer and visualize the gene regulatory network

Next, we load all necessary packages:
```{r}
suppressMessages(library(ArchR))
suppressMessages(library(Seurat))
suppressMessages(library(Signac))
suppressMessages(library(scMEGA))
suppressMessages(library(harmony))
suppressMessages(library(Nebulosa))
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(JASPAR2020))
suppressMessages(library(TFBSTools))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
suppressMessages(library(igraph))
suppressMessages(library(ggraph))
```

We first load the integrated data:
```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
coembed.sub <- readRDS("./Myofib/coembed.sub.rds")
```

## Trajectory analysis

### Dimensionality reduction
To infer trajectory, we will perform dimension reduction using diffusion map via the function _RunDiffusionMap_. This is based on the R package [destiny](https://bioconductor.org/packages/release/bioc/html/destiny.html).

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
coembed.sub <- RunDiffusionMap(coembed.sub, reduction = "harmony")

cols <- ArchR::paletteDiscrete(coembed.sub@meta.data[, "RNA_snn_res.0.2"])

p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.2", label = TRUE,
             reduction = "dm", shuffle = TRUE, cols = cols) +
    xlab("DC 1") + ylab("DC 2")

p
```

We can also plot snATAC-seq and snRNA-seq individually

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
DimPlot(coembed.sub, reduction = "dm", 
        group.by = "RNA_snn_res.0.2", split.by = "tech", cols = cols)
```

### Cell pairing
Next, we match the cells between these two modalities. In other words, for each cell in, for example, snATAC-seq, we will find a cell from snRNA-seq data so that these two cells have the similar profiles. This is only necessary when each modality was performed independently. If snRNA-seq and snATAC-seq data was generated by multi-modal protocol, e.g., 10X multiome or SHARE-seq, this step can be skipped.

We here use the method proposed by [Kartha, Vinay K., et al.](https://www.biorxiv.org/content/10.1101/2021.07.28.453784v1.abstract) to match the cells.

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
df.pair <- PairCells(object = coembed.sub, reduction = "harmony",
                    pair.by = "tech", ident1 = "ATAC", ident2 = "RNA")
```

We can visualize the paired cells

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
sel_cells <- c(df.pair$ATAC, df.pair$RNA)
coembed.sub2 <- coembed.sub[, sel_cells]

options(repr.plot.height = 5, repr.plot.width = 10)
DimPlot(coembed.sub2, reduction = "dm", 
        group.by = "RNA_snn_res.0.2", split.by = "tech", cols = cols)
```

We next create a new Seurat object for there paired cells as if they are generated by single-cell
multimodal protocol.

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
obj.pair <- CreatePairedObject(df.pair = df.pair, 
                               object = coembed.sub2,
                               use.assay1 = "RNA", 
                               use.assay2 = "ATAC")

obj.pair
```

Finally, we infer a pseudo-time trajectory from SCARA5+ fibroblasts to myofibroblast using the approach from [ArchR](https://www.archrproject.com/). Here we modified the function to allow to take a Seurat object as input

```{r, fig.height = 5, fig.width = 5, fig.align = "center"}
obj.pair <- AddTrajectory(object = obj.pair, 
                          trajectory = c(2, 1),
                          group.by = "RNA_snn_res.0.2", 
                          reduction = "dm",
                          dims = 1:3, 
                          use.all = FALSE)
                          
# we only plot the cells that are in this trajectory
obj <- obj.pair[, !is.na(obj.pair$Trajectory)]

p <- TrajectoryPlot(object = obj, 
                    reduction = "dm",
                    continuousSet = "blueYellow",
                    size = 1,
                   addArrow = FALSE)

p
```

## TF and gene selection
We next select candidate TFs and genes for building a meaningful gene regulatory network.

### Select TFs
To identify potential regulator (i.e., TFs), we first estimate an acitivty score for each TF in each cell. This is done by first performing motif matching and then computing deviation scores using [chromVAR](https://greenleaflab.github.io/chromVAR/index.html).

```{r}
# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
obj <- AddMotifs(
  object = obj,
  genome = BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm,
    assay = "ATAC"
)

obj <- RunChromVAR(
  object = obj,
  genome = BSgenome.Hsapiens.UCSC.hg38,
    assay = "ATAC"
)
```

We can visualize the TF activity dynamic along the trajectory

```{r, fig.height = 6, fig.width = 6, fig.align = "center"}
trajMM <- GetTrajectory(obj, 
                        assay = "chromvar", 
                        slot = "data",
                       smoothWindow = 7,
                       log2Norm = FALSE)

rownames(trajMM) <- obj@assays$ATAC@motifs@motif.names

ht <- TrajectoryHeatmap(trajMM, 
                        varCutOff = 0,
                        pal = paletteContinuous(set = "solarExtra"),
                        limits = c(-2, 2),
                       name = "TF activity")

ht
```

Apparently many TFs are showing dynamic TF activity along the trajectory. To reduce the number of candidates, we will use snRNA-seq to pick up the TFs that also show dynamic expression.

```{r}
trajRNA <- GetTrajectory(obj, 
                        assay = "RNA", 
                        slot = "data",
                       smoothWindow = 7,
                       log2Norm = TRUE)
                       
df.cor <- GetCorrelation(trajMM, trajRNA)

# we only select TFs that show significant correlation
df.cor <- df.cor[df.cor$adj_p < 0.01 & df.cor$correlation > 0.3, ]
```

We can try to assign a pseudo-time label for each TF.

```{r}
matMM <- TrajectoryHeatmap(trajMM, 
                        varCutOff = 0,
                        pal = paletteContinuous(set = "solarExtra"),
                        limits = c(-2, 2),
                       name = "TF activity",
                       returnMatrix = TRUE)

df_tf_time_point <- data.frame(tfs = rownames(matMM),
                               time_point = seq(1, 100, length.out = nrow(matMM)))
rownames(df_tf_time_point) <- df_tf_time_point$tfs

df_tf_time_point <- df_tf_time_point[df.cor$tfs, ]
df.cor$time_point <- df_tf_time_point$time_point
df.cor <- df.cor[order(df.cor$time_point), ]

head(df.cor)
```

We can visualize the TF activity and expression dynamics along the trajectory by generating a joint heatmap plot.

```{r, fig.height = 8, fig.width = 8, fig.align = "center"}
trajMM <- trajMM[df.cor$tfs, ]
trajRNA <- trajRNA[df.cor$tfs, ]

ht <- CorrelationHeatmap(trajectory1 = trajMM, 
                         trajectory2 = trajRNA,
                         name1 = "TF activity",
                         name2 = "Gene expression")

ht
```

### Select genes

We will select relevant genes by first linking genes to peaks based on the corrleation between gene expression from the snRNA-seq data and peak accessibility from the snATAC-seq data along the inferred trajectory. This means that we only consider a gene to be a potential target if it can be assocaited to at least one peak.

Let's first get the gene expression and chromatin accessibility data along our trajectory

```{r}
trajRNA <- GetTrajectory(obj, 
                        assay = "RNA", 
                        slot = "data",
                       smoothWindow = 7,
                       log2Norm = TRUE)

trajATAC <- GetTrajectory(obj, 
                        assay = "ATAC", 
                        slot = "data",
                       smoothWindow = 7,
                       log2Norm = TRUE)

# note here we only use the top 10% most variable genes
groupMatRNA <- TrajectoryHeatmap(trajRNA,
                        varCutOff = 0.9,
                        pal = paletteContinuous(set = "horizonExtra"),
                        limits = c(-2, 2),
                                 returnMatrix = TRUE)

groupMatATAC <- TrajectoryHeatmap(trajATAC, 
                        varCutOff = 0,
                                  maxFeatures = nrow(trajATAC),
                            pal = paletteContinuous(set = "solarExtra"),
                            limits = c(-2, 2),
                       name = "Chromatin accessibility",
                                 returnMatrix = TRUE)
```

The peak-to-gene link can be identified by using the function _PeakToGene_

```{r}
df.p2g <- PeakToGene(peak.mat = groupMatATAC,
                     gene.mat = groupMatRNA, 
                     genome = "hg38")
```

We here filter the links based on the following criteria:

* The peak must be a distal regulatory element which likely play a role as enhancer

* The correlation between gene expression and peak accessibility muse be positive and significant

```{r}
df.p2g <- df.p2g %>%
    subset(distance > 2000) %>%
    subset(Correlation > 0 & FDR < 1e-04)
```

After filtering, we can also generate a joint heatmap to visualize the peak accessibility and associated gene expression dynamics along the trajectory

```{r, fig.height = 8, fig.width = 12, fig.align = "center"}
trajATAC <- trajATAC[df.p2g$peak, ]
trajRNA <- trajRNA[df.p2g$gene, ]

ht <- CorrelationHeatmap(trajectory1 = trajATAC, 
                         trajectory2 = trajRNA,
                        name1 = "Chromatin accessibility",
                        name2 = "Gene expression",
                        labelRows1 = FALSE,
                        labelRows2 = FALSE)

ht
```

## Gene regulatory network inference and visualization

We here will try to predict a gene regulatory network based on the correlation of the TF binding activity as estimated from snATAC-seq and gene expression as measured by snRNA-seq along the trajectory.

```{r}
tf.gene.cor <- GetTFGeneCorrelation(object = obj, 
                                    tf.use = df.cor$tfs, 
                                    gene.use = unique(df.p2g$gene),
                                    tf.assay = "chromvar", 
                                    gene.assay = "RNA",
                                    trajectory.name = "Trajectory")
```

We can then visualize this correlation matrix by heatmap. Also, we can cluster the genes and TFs to identify different regulatory modules for the predefined sub-populations.

```{r, fig.height = 6, fig.width = 12, fig.align = "center"}
ht <- GRNHeatmap(tf.gene.cor, 
                 tf.timepoint = df.cor$time_point)
ht
```

To associate genes to TFs, we will use the peak-to-gene links and TF binding sites information. Specifically, if a gene is regulated by a peak and this peak is bound by a TF, then we consider this gene to be a target of this TF.

```{r, fig.height = 6, fig.width = 12, fig.align = "center"}
motif.matching <- obj@assays$ATAC@motifs@data
colnames(motif.matching) <- obj@assays$ATAC@motifs@motif.names
motif.matching <-
    motif.matching[unique(df.p2g$peak), unique(tf.gene.cor$tf)]


df.grn <- GetGRN(motif.matching = motif.matching, 
                 df.cor = tf.gene.cor, 
                 df.p2g = df.p2g)
```

Finally, we can visualize our network as the last step of this analysis

```{r viz_network, fig.height = 15, fig.width = 15, fig.align = "center"}
# define colors for nodes representing TFs (i.e., regulators)
df.cor <- df.cor[order(df.cor$time_point), ]
tfs.timepoint <- df.cor$time_point
names(tfs.timepoint) <- df.cor$tfs

# plot the graph, here we can highlight some genes
df.grn2 <- df.grn %>%
    subset(correlation > 0.4) %>%
    select(c(tf, gene, correlation)) %>%
    rename(weights = correlation)
    
p <- GRNPlot(df.grn2, 
             tfs.timepoint = tfs.timepoint,
             show.tf.labels = TRUE,
             seed = 42, 
             plot.importance = FALSE,
            min.importance = 2,
            remove.isolated = FALSE)

options(repr.plot.height = 20, repr.plot.width = 20)

print(p)
```

```{r}
# Check session information
sessionInfo()
```

