---
title: "Gene-regulatory network of myofibroblast differentiation in myocardial infarction"
output: html_document
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

During this tutorial, we will integrate the snRNA-seq and snATAC-seq data
generated from the human heart samples after myocardial infarction. The integrated
data will be used as input for inferring gene regulatory network.

We first download the required data. In this case, we need two Seurat objects 
with each one corresponding to snRNA-seq and snATAC-seq respectively. 
The snRNA-seq object includes gene expression data of all fibroblasts and 
the snATAC-seq includes all chromatin accessibility profiles. 
Additionally, we also need a gene activity matrix for data integration. 
This matrix was estimated from the snATAC-seq data by using 
the [ArchR](https://www.archrproject.com/) package. 
The script of cleaning the data and preparing these objects is found [here](https://costalab.ukaachen.de/open_data/scMEGA/Fibroblast/01_prepare_data.html).

Run the following commands to download the data:
```{r}
dir.create('./Fibroblast')

download.file(url = 'https://costalab.ukaachen.de/open_data/scMEGA/Fibroblast/snRNA.rds', 
              destfile = './Fibroblast/snRNA.rds', 
              method = 'wget', extra = '--no-check-certificate')

download.file(url = 'https://costalab.ukaachen.de/open_data/scMEGA/Fibroblast/snATAC.rds', 
              destfile = './Fibroblast/snATAC.rds', 
              method = 'wget', extra = '--no-check-certificate')
              
download.file(url = 'https://costalab.ukaachen.de/open_data/scMEGA/Fibroblast/gene.activity.rds', 
              destfile = './Fibroblast/gene.activity.rds', 
              method = 'wget', extra = '--no-check-certificate')
```


Next, we load all necessary packages:
```{r}
suppressMessages(library(ArchR))
suppressMessages(library(Seurat))
suppressMessages(library(Signac))
suppressMessages(library(scMEGA))
suppressMessages(library(harmony))
suppressMessages(library(Nebulosa))
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressMessages(library(JASPAR2020))
suppressMessages(library(TFBSTools))
suppressMessages(library(igraph))
suppressMessages(library(ggraph))
```

## Data integration

Let's load the data into memory and see how they look like

```{r}
obj.rna <- readRDS("./Fibroblast/snRNA.rds")
obj.atac <- readRDS("./Fibroblast/snATAC.rds")
gene.activity <- readRDS("./Fibroblast/gene.activity.rds")

obj.rna
obj.atac
```

We can observe that there are 45,515 and 6,481 cells in our snRNA-seq and snATAC-seq datasets. We now visualize the data as colored by patients. Note that here we used the UMAP embedding generated from batch-corrected low-dimensional space so that no batch effects are observed from the 2D visualization.


```{r, fig.height = 5, fig.width = 12, fig.align = "center"}
p1 <- DimPlot(obj.rna, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snRNA-seq")
    
p2 <- DimPlot(obj.atac, pt.size = 1, reduction = "umap_harmony") +
    ggtitle("snATAC-seq")

p1 + p2
```

### Co-embedding
First, we need to project the data into a common low-dimensional space. This is done by using the CCA method from [Seurat](https://satijalab.org/seurat/). To this end, we have wrapped several functions from Seurat into a single function _CoembedData_.


```{r}
obj.coembed <- CoembedData(
  obj.rna,
  obj.atac, 
  gene.activity, 
  weight.reduction = "harmony", 
  verbose = FALSE
)
```

We next visualize the snRNA-seq and snATAC-seq in this shared UMAP space. The cells are colored by
patients or modalities.

```{r, fig.height = 5, fig.width = 12, fig.align = "center"}

p1 <- DimPlot(obj.coembed, group.by = "patient", shuffle = TRUE, label = TRUE)
p2 <- DimPlot(obj.coembed, group.by = "tech", shuffle = TRUE, label = TRUE)

p1 + p2
```

The batch effects between patients, regions and modalities are quite clear. 
So next we use [Harmony](https://www.nature.com/articles/s41592-019-0619-0) to 
perform batch correction and generate a new UMAP embedding.

```{r, fig.height = 10, fig.width = 12, fig.align = "center"}
obj.coembed <- RunHarmony(
  obj.coembed,
  group.by.vars = c("patient", "region", "tech"),
  reduction = "pca",
  max.iter.harmony = 30,
  dims.use = 1:30,
  project.dim = FALSE,
  plot_convergence = FALSE
)

obj.coembed <- RunUMAP(
  obj.coembed,
  dims = 1:30,
  reduction = 'harmony',
  reduction.name = "umap_harmony",
  reduction.ke = 'umapharmony_',
  verbose = FALSE,
  min.dist = 0.4
)
```

We can plot the data again

```{r, fig.height = 5, fig.width = 12, fig.align = "center"}
p1 <-
  DimPlot(obj.coembed, group.by = "patient", reduction = "umap_harmony")
p2 <-
  DimPlot(obj.coembed, group.by = "tech", reduction = "umap_harmony")

p1 + p2
```

From the new UMAP embedding, we can observe that after batch-correction, cells from different patients, regions, and modalities are well mixed.

Based on our previous works of myofibroblast differentiation in [human](https://www.nature.com/articles/s41586-020-2941-1) and [mouse](https://www.nature.com/articles/s41467-021-26530-2) kidney, we already known some relevant genes for this biological process. For example, SCARA5 is a marker for myofibroblast progenitor, and COL1A1, POSTN, and FN1 are highly expressed in myofibroblast. Therefore we can visualize the expression of these genes to check if we can also identify similar process in human heart. Note that to make the visualization clear, here we used the package [Nebulosa](https://github.com/powellgenomicslab/Nebulosa) to plot the data.

```{r, fig.height = 10, fig.width = 12, fig.align = "center"}
p1 <-
  plot_density(obj.coembed,
               features = "SCARA5",
               reduction = "umap_harmony",
               pal = "magma")
p2 <-
  plot_density(obj.coembed,
               features = "COL1A1",
               reduction = "umap_harmony",
               pal = "magma")
p3 <-
  plot_density(obj.coembed,
               features = "POSTN",
               reduction = "umap_harmony",
               pal = "magma")
p4 <-
  plot_density(obj.coembed,
               features = "FN1",
               reduction = "umap_harmony",
               pal = "magma")


(p1 + p2) / (p3 + p4)
```

From the visualization, we can observe that some cells highly express SCARA5 which could be the progenitors of myofibroblasts. On the other hand, some cells highly express COL1A1, POSTN, and FN1 and they could be terminally differentiated myofibroblasts.

### Sub-clustering

We next perform sub-clustering to identify different populations in our multi-omic fibroblast data. To further control the data quality, here we will use a two-round approach to remove low-quality cells. We first use a high-resolution to get a large amount of clusters.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
obj.coembed <- FindNeighbors(obj.coembed, reduction = "harmony", dims = 1:30)
obj.coembed <- FindClusters(obj.coembed, resolution = 0.9, verbose = FALSE)

cols <- ArchR::paletteDiscrete(obj.coembed@meta.data[, "RNA_snn_res.0.9"])
    
p <- DimPlot(obj.coembed, group.by = "RNA_snn_res.0.9", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE) +
    scale_color_manual(values = cols) +
    xlab("UMAP1") + ylab("UMAP2")
    
p
```

We can use the function _CellPropPlot_ to visualize the cell propotion across all patients.

```{r, fig.height = 6, fig.width = 10, fig.align = "center"}
p <- CellPropPlot(obj.coembed,
                  group.by = "RNA_snn_res.0.9",
                  prop.in = "patient_region_id",
                  cols = cols)

p
```

Next, we identify the markers for each cluster and visualize the top 3.

```{r, fig.height = 5, fig.width = 15, fig.align = "center", eval=TRUE}
all.markers <- FindAllMarkers(obj.coembed, 
                              only.pos = TRUE, 
                              min.pct = 0.25, logfc.threshold = 0.5)

df <- all.markers %>%
    group_by(cluster) %>%
    slice_max(n = 3, order_by = avg_log2FC)

p <- DotPlot(obj.coembed, features = unique(df$gene)) + RotatedAxis()

print(p)

saveRDS(obj.coembed, "./Fibroblast/coembed.rds")
```

The above dot plot demonstrates the top 3 markers per cluster and we can easily classify cluster
3 and 7 as myofibroblasts. In addition, two clusters (i.e., cluster 2 and 5) are marked by RYR2, a well known marker gene for cardiomyocyte, likely due to the background noise. It is worth pointing out that in another study [Cells of the adult human heart](https://www.nature.com/articles/s41586-020-2797-4) a similar fibroblast sub-population (marked by another cardiomyocyte-specific gene TNNT2) was also identified. Based on these, we decide to remove these two clusters.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
Idents(obj.coembed) <- "RNA_snn_res.0.9"
coembed.sub <- subset(obj.coembed, idents = c(2, 5), invert = TRUE)
coembed.sub

cols.clusters <- ArchR::paletteDiscrete(coembed.sub@meta.data[, "RNA_snn_res.0.9"])

p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.9", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE, cols = cols) +
    xlab("UMAP1") + ylab("UMAP2")

p
```

We then re-do the UMAP embedding and clustering with a lower resolution to reduce complexity.

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
coembed.sub <- RunUMAP(coembed.sub, 
                       dims = 1:30, 
                       reduction = 'harmony',
                       reduction.name = "umap_harmony",
                       reduction.key = 'umap_harmony_',
                       verbose = FALSE,
                       min.dist = 0.4)

## re-clustering
coembed.sub <- FindNeighbors(coembed.sub, reduction = "harmony", dims = 1:30)
coembed.sub <- FindClusters(coembed.sub, resolution = 0.2, verbose = FALSE)

cols <- ArchR::paletteDiscrete(coembed.sub@meta.data[, "RNA_snn_res.0.2"])
    
p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.2", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE) +
    scale_color_manual(values = cols) +
    xlab("UMAP1") + ylab("UMAP2")
    
p                
```

Marker genes are identified based on new clustering results and we can plot the
top 10 markers.

```{r, fig.height = 5, fig.width = 15, fig.align = "center"}
all.markers <- FindAllMarkers(coembed.sub, 
                              only.pos = TRUE, 
                              min.pct = 0.25, logfc.threshold = 0.5)

df <- all.markers %>%
    group_by(cluster) %>%
    slice_max(n = 10, order_by = avg_log2FC)

p <- DotPlot(coembed.sub, features = unique(df$gene)) + RotatedAxis()

print(p)
```

Now the clustering results are clearer, and cluster 1 is myofibroblast, and cluster 2 is SCARA5+ fibroblast.

We can plot the snRNA-seq and snATAC-seq data separately

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.2", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE, split.by = "tech", 
             cols = cols) +
    xlab("UMAP1") + ylab("UMAP2")

p
```

Visualize the cell proportion of each sub-population across all patients

```{r, fig.height = 6, fig.width = 10, fig.align = "center"}
p <- CellPropPlot(coembed.sub, 
                   group.by = "RNA_snn_res.0.2", 
                   prop.in = "patient_region_id", 
                   cols = cols)

p
```

Since we have annotated all patients into three major groups, i.e., myogenic, ischmeic, and fibrotic.
we can also perform statistical test to check if any sub-population are enriched in any of the above group. This can be done by the function _CompareCellProp_.

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
coembed.sub$patient_group <- factor(coembed.sub$patient_group, 
                                    levels = c("myogenic", "ischemic", "fibrotic"))

p <- CompareCellProp(object = coembed.sub, 
                     group.by = "RNA_snn_res.0.2", 
                     prop.in = "patient_region_id", 
                      sample.annotation = "patient_group",
                    comparisons = list(c("myogenic", "ischemic"),
                                       c("ischemic", "fibrotic"),
                                       c("myogenic", "fibrotic")))

p

saveRDS(coembed.sub, "./Fibroblast/coembed.cluster.rds")
```

This analysis reveals significant changes of cell proportion between different conditions for cluster 1, 2, and 3. We therefore only use these three clusters in our trajectory analysis.

```{r}
Idents(coembed.sub) <- "RNA_snn_res.0.2"
coembed.sub <- subset(coembed.sub, idents = c(1, 2, 3))
```

We can generate a new UMAP

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
cols <- ArchR::paletteDiscrete(unique(coembed.sub@meta.data[, "RNA_snn_res.0.2"]))

coembed.sub <- RunUMAP(coembed.sub, 
               dims = 1:30, 
               reduction = 'harmony',
               reduction.name = "umap_harmony",
               reduction.ke = 'umap_harmony_',
              verbose = FALSE)

p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.2", label = TRUE,
             reduction = "umap_harmony", shuffle = TRUE, cols = cols) +
    xlab("UMAP1") + ylab("UMAP2")

p
```


Save data
```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
saveRDS(coembed.sub, "./Fibroblast/coembed.sub.rds")
```

## Trajectory analysis
We next identify the trajectory for myofibroblast differentiation.

### Dimensionality reduction
To infer trajectory, we will perform dimension reduction using diffusion map via the function _RunDiffusionMap_. This is based on the R package [destiny](https://bioconductor.org/packages/release/bioc/html/destiny.html).

```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
coembed.sub <- RunDiffusionMap(coembed.sub, reduction = "harmony")

cols <- ArchR::paletteDiscrete(coembed.sub@meta.data[, "RNA_snn_res.0.2"])

p <- DimPlot(coembed.sub, group.by = "RNA_snn_res.0.2", label = TRUE,
             reduction = "dm", shuffle = TRUE, cols = cols) +
    xlab("DC 1") + ylab("DC 2")

p
```

We can also plot snATAC-seq and snRNA-seq individually

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
DimPlot(coembed.sub, reduction = "dm", 
        group.by = "RNA_snn_res.0.2", split.by = "tech", cols = cols)
```

### Cell pairing
Next, we match the cells between these two modalities. In other words, for each cell in, for example, snATAC-seq, we will find a cell from snRNA-seq data so that these two cells have the similar profiles. This is only necessary when each modality was performed independently. If snRNA-seq and snATAC-seq data was generated by multi-modal protocol, e.g., 10X multiome or SHARE-seq, this step can be skipped.

We here use the method proposed by [Kartha, Vinay K., et al.](https://www.sciencedirect.com/science/article/pii/S2666979X22001082?via%3Dihub) to match the cells.

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
df.pair <- PairCells(object = coembed.sub, reduction = "harmony",
                    pair.by = "tech", ident1 = "ATAC", ident2 = "RNA")
```

We can visualize the paired cells

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
sel_cells <- c(df.pair$ATAC, df.pair$RNA)
coembed.sub2 <- coembed.sub[, sel_cells]

options(repr.plot.height = 5, repr.plot.width = 10)
DimPlot(coembed.sub2, reduction = "dm", 
        group.by = "RNA_snn_res.0.2", split.by = "tech", cols = cols)
```

We next create a new Seurat object for there paired cells as if they are generated by single-cell
multimodal protocol.

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
obj.pair <- CreatePairedObject(df.pair = df.pair, 
                               object = coembed.sub2,
                               use.assay1 = "RNA", 
                               use.assay2 = "ATAC")

obj.pair
```

Finally, we infer a pseudo-time trajectory from SCARA5+ fibroblasts to myofibroblast using the approach from [ArchR](https://www.archrproject.com/). Here we modified the function to allow to take a Seurat object as input

```{r, fig.height = 5, fig.width = 5, fig.align = "center"}
obj.pair <- AddTrajectory(object = obj.pair, 
                          trajectory = c(2, 1),
                          group.by = "RNA_snn_res.0.2", 
                          reduction = "dm",
                          dims = 1:3, 
                          use.all = FALSE)
                          
# we only plot the cells that are in this trajectory
obj <- obj.pair[, !is.na(obj.pair$Trajectory)]

p <- TrajectoryPlot(object = obj, 
                    reduction = "dm",
                    continuousSet = "blueYellow",
                    size = 1,
                   addArrow = FALSE)

p
```

## TF and gene selection
We next select candidate TFs and genes for building a meaningful gene regulatory network.

### Select TFs
To identify potential regulator (i.e., TFs), we first estimate an acitivty score for each TF in each cell. This is done by first performing motif matching and then computing deviation scores using [chromVAR](https://greenleaflab.github.io/chromVAR/index.html).

```{r}
# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
obj <- AddMotifs(
  object = obj,
  genome = BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm,
    assay = "ATAC"
)

obj <- RunChromVAR(
  object = obj,
  genome = BSgenome.Hsapiens.UCSC.hg38,
    assay = "ATAC"
)

res <- SelectTFs(object = obj, return.heatmap = TRUE)

df.cor <- res$tfs
ht <- res$heatmap
```

We can visualize the TF activity dynamic along the trajectory

```{r, fig.height = 6, fig.width = 8, fig.align = "center"}
draw(ht)
```

### Select genes

We will select relevant genes by first linking genes to peaks based on the corrleation between gene expression from the snRNA-seq data and peak accessibility from the snATAC-seq data along the inferred trajectory. This means that we only consider a gene to be a potential target if it can be assocaited to at least one peak.
```{r, fig.height = 6, fig.width = 8, fig.align = "center"}
res <- SelectGenes(object = obj,
                  labelTop1 = 0,
                  labelTop2 = 0)

df.p2g <- res$p2g
ht <- res$heatmap

draw(ht)
```

## Gene regulatory network inference

We here will try to predict a gene regulatory network based on the correlation of the TF binding activity as estimated from snATAC-seq and gene expression as measured by snRNA-seq along the trajectory.

```{r}
tf.gene.cor <- GetTFGeneCorrelation(object = obj, 
                                    tf.use = df.cor$tfs, 
                                    gene.use = unique(df.p2g$gene),
                                    tf.assay = "chromvar", 
                                    gene.assay = "RNA",
                                    trajectory.name = "Trajectory")
```

We can then visualize this correlation matrix by heatmap. Also, we can cluster the genes and TFs to identify different regulatory modules for the predefined sub-populations.

```{r, fig.height = 6, fig.width = 12, fig.align = "center"}
ht <- GRNHeatmap(tf.gene.cor, 
                 tf.timepoint = df.cor$time_point)
ht
```

To associate genes to TFs, we will use the peak-to-gene links and TF binding sites information. Specifically, if a gene is regulated by a peak and this peak is bound by a TF, then we consider this gene to be a target of this TF.

```{r, fig.height = 6, fig.width = 12, fig.align = "center"}
motif.matching <- obj@assays$ATAC@motifs@data
colnames(motif.matching) <- obj@assays$ATAC@motifs@motif.names
motif.matching <-
    motif.matching[unique(df.p2g$peak), unique(tf.gene.cor$tf)]


df.grn <- GetGRN(motif.matching = motif.matching, 
                 df.cor = tf.gene.cor, 
                 df.p2g = df.p2g)
```

Next, we can visualize our network as the last step of this analysis

```{r viz_network, fig.height = 15, fig.width = 15, fig.align = "center"}
# define colors for nodes representing TFs (i.e., regulators)
df.cor <- df.cor[order(df.cor$time_point), ]
tfs.timepoint <- df.cor$time_point
names(tfs.timepoint) <- df.cor$tfs

# plot the graph, here we can highlight some genes
df.grn2 <- df.grn %>%
    subset(correlation > 0.4) %>%
    select(c(tf, gene, correlation)) %>%
    rename(weights = correlation)

saveRDS(df.grn2,"./Fibroblast/final_grn.Rds")    

p <- GRNPlot(df.grn2, 
             tfs.timepoint = tfs.timepoint,
             show.tf.labels = TRUE,
             seed = 42, 
             plot.importance = FALSE,
            min.importance = 2,
            remove.isolated = FALSE)

options(repr.plot.height = 20, repr.plot.width = 20)

print(p)
```

## GRN visualization

### GRN along the trajectory

Once we generated the gene regulatory network, we can visualize individual TFs in terms of binding activity, expression, and target expression along the pseudotime trajectory.

Here we select two TFs for visualization.

```{r, fig.height = 3.5, fig.width = 12, fig.align = "center"}
obj <- AddTargetAssay(object = obj, df.grn = df.grn2)

p1 <- PseudotimePlot(object = obj, tf.use = "NR3C2")
p2 <- PseudotimePlot(object = obj, tf.use = "RUNX1")

p1 + p2
```

The x-axis in above plots present pseudotime point along the trajectory, and the y-axis represent TF binding acitivty, TF expression, and TF target expression after z-score transformation.

### GRN in space

As an additional step, if the spatial transcriptome (ST) data (e.g., generated by using 10X Visium assay) are available, we can then visualize the target gene expression for some interesting TFs to understand the regulon activity in spatial space. First let's download the ST data

```{bash, eval=TRUE}
dir.create('./VisiumSpatial')


download.file(url = 'https://costalab.ukaachen.de/open_data/scMEGA/VisiumSpatial/AKK004_157772.rds', 
              destfile = './VisiumSpatial/AKK004_157772.rds', 
              method = 'wget', extra = '--no-check-certificate')
```

Of note, each spot in the 10X Visium assay contains multiple cells, meaning that we usually cannot identify cell types by clustering the data. To address this issue, people developed various computation tools to estimate cell-type-specific abundance in each spot. Here we have perform such analysis using [cell2location](https://www.nature.com/articles/s41587-021-01139-4) based on our snRNA-seq data. Let's first visualize the fibroblast proportion.


```{r, fig.height = 5, fig.width = 5, fig.align = "center"}
obj.spatial <- readRDS("./VisiumSpatial/AKK004_157772.rds")

DefaultAssay(obj.spatial) <- "c2l_props"

p <- SpatialFeaturePlot(obj.spatial, features = "Fib", min.cutoff = "q5",
                       max.cutoff = "q95") +
    scale_fill_viridis(option = "D") +
    ggtitle("") +
    labs(fill = "") +
    theme(legend.position = "bottom")

print(p)
```

We next can check the gene expression in space. Here, we use the above TFs.


```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
DefaultAssay(obj.spatial) <- "SCT"

p1 <- SpatialFeaturePlot(object = obj.spatial, features = "NR3C2") +
    ggtitle("") +
    theme(legend.position = "bottom") +
scale_fill_viridis(option = "C")

p2 <- SpatialFeaturePlot(object = obj.spatial, features = "RUNX1") +
    ggtitle("") +
    theme(legend.position = "bottom") +
scale_fill_viridis(option = "C")

p1 + p2
```

Then, we visualize the regulon activity of NR3C2 and RUNX1 by using the GRN that we predicted. The major idea is that if the target genes of a TF are up-regulated, then this TF should have high activity. This principle has been used to develop computation methods to estimate TF activity from gene expression data, such as [DoRothEA](https://genome.cshlp.org/content/29/8/1363.short).

```{r, fig.height = 5, fig.width = 10, fig.align = "center"}
DefaultAssay(obj.spatial) <- "SCT"

p1 <- GRNSpatialPlot(object = obj.spatial, assay = "SCT", df.grn = df.grn2,
                   tf.use = "NR3C2", vis.option = "B") +
    ggtitle("") +
    theme(legend.position = "bottom") 

p2 <- GRNSpatialPlot(object = obj.spatial, assay = "SCT", df.grn = df.grn2,
                   tf.use = "RUNX1", vis.option = "B") +
    ggtitle("") +
    theme(legend.position = "bottom")

p1 + p2
```

The above plots demonstrate a clear spatial-distributed pattern for the regulon activity of NR3C2 and RUNX1.

Save data
```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
saveRDS(obj, "./Fibroblast/coembed.trajectory.rds")
```

## Network analysis

As an addition step, we next will explore the gene regulatory networks generated by scMEGA in fibroblasts during myocardial infarction by using network analysis methods. 

This will allow us to 

* explore regulators regarding their network topological properties
* to vizualise regulatory networks centred in selected transcription factors.  

We first load the network produced by scMEGA:
```{r, fig.height = 5, fig.width = 6, fig.align = "center"}
dfgrn <- readRDS("./Fibroblast/final_grn.Rds")
netobj <- graph_from_data_frame(dfgrn,directed = TRUE)
V(netobj)$type <- ifelse(V(netobj)$name %in% dfgrn$tf,"TF/Gene","Gene")
```

Next, we will estimate distint network topological measures as node degree, node centrality and node page rank score and plot these in a principal component based embedding with a focus on transcription factors. We plot bellow the 3 first PCs. 

### Network topological measures embedding: PC1 and PC2
```{r,fig.width=12,fig.height=12}
p <- TopEmbGRN(df.grn=netobj)
```

### Network topological measures embedding: PC2 and PC3
```{r,fig.width=12,fig.height=12}
p <- TopEmbGRN(df.grn=netobj,axis=c(2,3))
```

This reveals that RUNX1 is a central TF in this network, as it has high in/out-degree values and high mediator score. 

### RUNX1 centric network. 

To further inspect genes and TFs associated with RUNX1, we create a centric layout embedding.

```{r,fig.width=12,fig.height=12}
NetCentPlot(netobj,"RUNX1")
```

This network reveals direct targets of RUNX1 such as TEAD4 and HIF1A and indirect targets as TEAD1 and TEAD2. If you want, you can explore a highlight parameter to personalize the plot by only showing the desired genes/TFs. By default, the TFs with a betweenness score higher than an average is depicted in the plot.


```{r}
# Check session information
sessionInfo()
```

