<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene-regulatory network of myofibroblast differentiation in myocardial infarction • scMEGA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene-regulatory network of myofibroblast differentiation in myocardial infarction">
<meta property="og:description" content="scMEGA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMEGA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/GRN-fibroblast.html">Infer GRN for myofibroblast</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/CostaLab/scMEGA" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene-regulatory network of myofibroblast differentiation in myocardial infarction</h1>
            
            <h4 data-toc-skip class="date">Compiled: May 11, 2022</h4>
      
      
      <div class="hidden name"><code>GRN-fibroblast.Rmd</code></div>

    </div>

    
    
<p>During this tutorial, we will be analyzing the snRNA-seq and snATAC-seq generated from human hearts after myocardial infarction to understand the gene regulatory dynamics for myofibroblast differentiation. More specifically, we will:</p>
<ul>
<li>Pair the cells between gene expression and chromatin accessibility data</li>
<li>Identify the trajectory for myofibroblast differentiation</li>
<li>Select the relevant TFs and genes for this process</li>
<li>Infer and visualize the gene regulatory network</li>
</ul>
<p>We first download all the required data. In this case, we need two Seurat objects with each one corresponding to snRNA-seq and snATAC-seq respectively. The snRNA-seq object includes gene expression data of all fibroblasts and snRNA-seq includes all chromatin accessibility data. Additionally, we also need a gene activity matrix for data integration. This matrix was estimated from the snATAC-seq data by using <a href="https://www.archrproject.com/" class="external-link">ArchR</a> package. The script of cleaning the data and preparing these objects is found <a href="https://www.dropbox.com/s/9i1lni0iia71ape/01_prepare_data.html?dl=0" class="external-link">here</a>.</p>
<p>Run the following commands to download the data:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://www.dropbox.com/s/6v98iiq8dtlcjho/snRNA.rds</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://www.dropbox.com/s/h30ej52burtsgdy/snATAC.rds</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://www.dropbox.com/s/7uc2ad0jcrzks20/gene.activity.rds</span></code></pre></div>
<p>Next, we load all necessary packages:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/timoast/signac" class="external-link">Signac</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://costalab.github.io/scMEGA/">scMEGA</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">harmony</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/powellgenomicslab/Nebulosa" class="external-link">Nebulosa</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Let’s load the data into memory and see how they look like</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./snRNA.rds"</span><span class="op">)</span>
<span class="va">obj.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./snATAC.rds"</span><span class="op">)</span>
<span class="va">gene.activity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./gene.activity.rds"</span><span class="op">)</span>

<span class="va">obj.rna</span></code></pre></div>
<pre><code><span class="co">## An object of class Seurat </span>
<span class="co">## 28933 features across 45515 samples within 1 assay </span>
<span class="co">## Active assay: RNA (28933 features, 2000 variable features)</span>
<span class="co">##  4 dimensional reductions calculated: pca, harmony, umap, umap_harmony</span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.atac</span></code></pre></div>
<pre><code><span class="co">## An object of class Seurat </span>
<span class="co">## 173843 features across 6481 samples within 1 assay </span>
<span class="co">## Active assay: ATAC (173843 features, 0 variable features)</span>
<span class="co">##  3 dimensional reductions calculated: harmony, umap, umap_harmony</span></code></pre>
<p>We can observe that there are 45,515 and 6,481 cells in our snRNA-seq and snATAC-seq datasets. We now visualize the data as colored by patients. Note that here we used the UMAP embedding generated from batch-corrected low-dimensional space so that no batch effects are observed from the 2D visualization.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.rna</span>, pt.size <span class="op">=</span> <span class="fl">1</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"snRNA-seq"</span><span class="op">)</span>
    
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.atac</span>, pt.size <span class="op">=</span> <span class="fl">1</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"snATAC-seq"</span><span class="op">)</span>

<span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-4-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="cell-pairing">Cell pairing<a class="anchor" aria-label="anchor" href="#cell-pairing"></a>
</h2>
<p>In this step, we match the cells between these two modalities. In other words, for each cell in, for example, snATAC-seq, we will find a cell from snRNA-seq data so that these two cells have the similar profiles.</p>
<div class="section level3">
<h3 id="co-embedding">Co-embedding<a class="anchor" aria-label="anchor" href="#co-embedding"></a>
</h3>
<p>First, we need to project the data into a common low-dimensional space. This is done by using the CCA method from <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>. More specifically, we wrapped several functions from Seurat into a single function <em>CoembedData</em>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoembedData.html">CoembedData</a></span><span class="op">(</span><span class="va">obj.rna</span>, <span class="va">obj.atac</span>, <span class="va">gene.activity</span>, 
                           weight.reduction <span class="op">=</span> <span class="st">"harmony"</span>, 
                           verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Performing data integration using Seurat...</span></code></pre>
<pre><code><span class="co">## Warning in RunCCA.Seurat(object1 = reference, object2 = query, features =</span>
<span class="co">## features, : Running CCA on different assays</span></code></pre>
<pre><code><span class="co">## Finding integration vectors</span></code></pre>
<pre><code><span class="co">## Finding integration vector weights</span></code></pre>
<pre><code><span class="co">## Transfering 19026 features onto reference data</span></code></pre>
<pre><code><span class="co">## Coemebdding the data...</span></code></pre>
<pre><code><span class="co">## Centering data matrix</span></code></pre>
<pre><code><span class="co">## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
<span class="co">## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span>
<span class="co">## This message will be shown once per session</span></code></pre>
<p>We next visualize the snRNA-seq and snATAC-seq in this shared space. The cells are coloded by patient, region, modality, or patient group.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"patient"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"region"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"tech"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"patient_group"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">p3</span> <span class="op">+</span> <span class="va">p4</span><span class="op">)</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-6-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>The batch effects between patients, regions and modalities are quite clear. So next we use <a href="https://www.nature.com/articles/s41592-019-0619-0" class="external-link">Harmony</a> to perform batch correction and generate a new UMAP embedding.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/harmony/man/RunHarmony.html" class="external-link">RunHarmony</a></span><span class="op">(</span><span class="va">obj.coembed</span>, 
                      group.by.vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"patient"</span>, <span class="st">"region"</span>, <span class="st">"tech"</span><span class="op">)</span>,
                     reduction <span class="op">=</span> <span class="st">"pca"</span>, 
                      max.iter.harmony <span class="op">=</span> <span class="fl">30</span>, 
                      dims.use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,
                     project.dim <span class="op">=</span> <span class="cn">FALSE</span>,
                     plot_convergence <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: Quick-TRANSfer stage steps exceeded maximum (= 2599800)</span></code></pre>
<pre><code><span class="co">## Harmony 1/30</span></code></pre>
<pre><code><span class="co">## Harmony 2/30</span></code></pre>
<pre><code><span class="co">## Harmony 3/30</span></code></pre>
<pre><code><span class="co">## Harmony 4/30</span></code></pre>
<pre><code><span class="co">## Harmony 5/30</span></code></pre>
<pre><code><span class="co">## Harmony 6/30</span></code></pre>
<pre><code><span class="co">## Harmony 7/30</span></code></pre>
<pre><code><span class="co">## Harmony 8/30</span></code></pre>
<pre><code><span class="co">## Harmony 9/30</span></code></pre>
<pre><code><span class="co">## Harmony 10/30</span></code></pre>
<pre><code><span class="co">## Harmony 11/30</span></code></pre>
<pre><code><span class="co">## Harmony 12/30</span></code></pre>
<pre><code><span class="co">## Harmony 13/30</span></code></pre>
<pre><code><span class="co">## Harmony 14/30</span></code></pre>
<pre><code><span class="co">## Harmony 15/30</span></code></pre>
<pre><code><span class="co">## Harmony 16/30</span></code></pre>
<pre><code><span class="co">## Harmony 17/30</span></code></pre>
<pre><code><span class="co">## Harmony 18/30</span></code></pre>
<pre><code><span class="co">## Harmony converged after 18 iterations</span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">obj.coembed</span>, 
               dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, 
               reduction <span class="op">=</span> <span class="st">'harmony'</span>,
               reduction.name <span class="op">=</span> <span class="st">"umap_harmony"</span>,
               reduction.ke <span class="op">=</span> <span class="st">'umapharmony_'</span>,
              verbose <span class="op">=</span> <span class="cn">FALSE</span>,
                   min.dist <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></code></pre></div>
<p>We can plot the data again</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"patient"</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"region"</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span>
<span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"tech"</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span>
<span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"patient_group"</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span>

<span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">p3</span> <span class="op">+</span> <span class="va">p4</span><span class="op">)</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-8-1.png" width="1152" style="display: block; margin: auto;"> The new UMAP embedding shows that after batch-correction, cells from different patients, regions, and modalities are well mixed.</p>
<p>Based on our previous works of myofibroblast differentiation in <a href="https://www.nature.com/articles/s41586-020-2941-1" class="external-link">human</a> and <a href="https://www.nature.com/articles/s41467-021-26530-2" class="external-link">mouse</a> kidney, we already known some relevant genes for this biological process. For example, SCARA5 is a marker for myofibroblast progenitor, and COL1A1, POSTN, and FN1 are highly expressed in myofibroblast. Therefore we can visualize the expression of these genes to check if we can also identify similar process in human heart. Note that to make the visualization clear, here we used the package <a href="https://github.com/powellgenomicslab/Nebulosa" class="external-link">Nebulosa</a> to plot the data.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Nebulosa/man/plot_density.html" class="external-link">plot_density</a></span><span class="op">(</span><span class="va">obj.coembed</span>, features<span class="op">=</span><span class="st">"SCARA5"</span>, reduction<span class="op">=</span><span class="st">"umap_harmony"</span>, pal <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Nebulosa/man/plot_density.html" class="external-link">plot_density</a></span><span class="op">(</span><span class="va">obj.coembed</span>, features<span class="op">=</span><span class="st">"COL1A1"</span>, reduction<span class="op">=</span><span class="st">"umap_harmony"</span>, pal <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span>
<span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Nebulosa/man/plot_density.html" class="external-link">plot_density</a></span><span class="op">(</span><span class="va">obj.coembed</span>, features<span class="op">=</span><span class="st">"POSTN"</span>, reduction<span class="op">=</span><span class="st">"umap_harmony"</span>, pal <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span>
<span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Nebulosa/man/plot_density.html" class="external-link">plot_density</a></span><span class="op">(</span><span class="va">obj.coembed</span>, features<span class="op">=</span><span class="st">"FN1"</span>, reduction<span class="op">=</span><span class="st">"umap_harmony"</span>, pal <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span>


<span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">p3</span> <span class="op">+</span> <span class="va">p4</span><span class="op">)</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-9-1.png" width="1152" style="display: block; margin: auto;"> From the visulization, we can observe that some cells highly express SCARA5 which could be the progentiors of myofibroblasts. On the other hand, some cells highly express COL1A1, POSTN, and FN1 and they could be terminally differentiated myofibroblasts.</p>
</div>
<div class="section level3">
<h3 id="sub-clustering-and-cleaning">Sub-clustering and cleaning<a class="anchor" aria-label="anchor" href="#sub-clustering-and-cleaning"></a>
</h3>
<p>We next perform sub-clustering to identify different populations in our fibroblast data. To further control the data quality, here we will use a two-round approach to remove low-quality cells. We first use a high-resolution to get a large amount of clusters.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">obj.coembed</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Computing nearest neighbor graph</span></code></pre>
<pre><code><span class="co">## Computing SNN</span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">obj.coembed</span>, resolution <span class="op">=</span> <span class="fl">0.9</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">ArchR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ArchR/man/paletteDiscrete.html" class="external-link">paletteDiscrete</a></span><span class="op">(</span><span class="va">obj.coembed</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="st">"RNA_snn_res.0.9"</span><span class="op">]</span><span class="op">)</span>
    
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.9"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,
             reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"UMAP1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"UMAP2"</span><span class="op">)</span>
    
<span class="va">p</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-10-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can use the function <em>CellPropPlot</em> to visualize the cell propotion across all patients.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CellPropPlot.html">CellPropPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, 
                   group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.9"</span>, 
                   prop.in <span class="op">=</span> <span class="st">"patient_region_id"</span>, 
                   cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span>

<span class="va">p</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-11-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Next, we identify the markers for each cluster and visualize the top 3.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">all.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">obj.coembed</span>, 
                              only.pos <span class="op">=</span> <span class="cn">TRUE</span>, 
                              min.pct <span class="op">=</span> <span class="fl">0.25</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Calculating cluster 0</span></code></pre>
<pre><code><span class="co">## Calculating cluster 1</span></code></pre>
<pre><code><span class="co">## Calculating cluster 2</span></code></pre>
<pre><code><span class="co">## Calculating cluster 3</span></code></pre>
<pre><code><span class="co">## Calculating cluster 4</span></code></pre>
<pre><code><span class="co">## Calculating cluster 5</span></code></pre>
<pre><code><span class="co">## Calculating cluster 6</span></code></pre>
<pre><code><span class="co">## Calculating cluster 7</span></code></pre>
<pre><code><span class="co">## Calculating cluster 8</span></code></pre>
<pre><code><span class="co">## Calculating cluster 9</span></code></pre>
<pre><code><span class="co">## Calculating cluster 10</span></code></pre>
<pre><code><span class="co">## Calculating cluster 11</span></code></pre>
<pre><code><span class="co">## Calculating cluster 12</span></code></pre>
<pre><code><span class="co">## Calculating cluster 13</span></code></pre>
<pre><code><span class="co">## Calculating cluster 14</span></code></pre>
<pre><code><span class="co">## Calculating cluster 15</span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">all.markers</span> <span class="op"><a href="https://rdrr.io/pkg/harmony/man/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/harmony/man/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span>, order_by <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-12-1.png" width="1440" style="display: block; margin: auto;"> The above dot plot demonstrates the top 3 markers per cluster and we can easily classify cluster 3 and 7 as myofibroblasts. In addtion, two clusters (i.e., cluster 2 and 5) are markerd by RYR2, a well known marker gene for cardiomyocyte, likely due to the background noise. It is worth pointing out that in another study <a href="https://www.nature.com/articles/s41586-020-2797-4" class="external-link">Cells of the adult human heart</a> a simialr fibroblast sub-population (marked by another cardiomyocyte-specific gene TNNT2) was also identified. Based on these, we decide to remove these two clusters.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">obj.coembed</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA_snn_res.0.9"</span>
<span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/signac/reference/subset.Motif.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">obj.coembed</span>, idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">5</span><span class="op">)</span>, invert <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">coembed.sub</span></code></pre></div>
<pre><code><span class="co">## An object of class Seurat </span>
<span class="co">## 221802 features across 43276 samples within 3 assays </span>
<span class="co">## Active assay: RNA (28933 features, 2000 variable features)</span>
<span class="co">##  2 other assays present: ATAC, GeneActivity</span>
<span class="co">##  4 dimensional reductions calculated: pca, umap, harmony, umap_harmony</span></code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols.clusters</span> <span class="op">&lt;-</span> <span class="fu">ArchR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ArchR/man/paletteDiscrete.html" class="external-link">paletteDiscrete</a></span><span class="op">(</span><span class="va">coembed.sub</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="st">"RNA_snn_res.0.9"</span><span class="op">]</span><span class="op">)</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.9"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,
             reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"UMAP1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"UMAP2"</span><span class="op">)</span>

<span class="va">p</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We then re-do the UMAP embedding and clustering with a lower resolution.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">coembed.sub</span>, 
                       dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, 
                       reduction <span class="op">=</span> <span class="st">'harmony'</span>,
                       reduction.name <span class="op">=</span> <span class="st">"umap_harmony"</span>,
                       reduction.key <span class="op">=</span> <span class="st">'umap_harmony_'</span>,
                       verbose <span class="op">=</span> <span class="cn">FALSE</span>,
                       min.dist <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: Keys should be one or more alphanumeric characters followed by an</span>
<span class="co">## underscore, setting key from umap_harmony_ to umapharmony_</span></code></pre>
<pre><code><span class="co">## Warning: All keys should be one or more alphanumeric characters followed by an</span>
<span class="co">## underscore '_', setting key to umapharmony_</span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## re-clustering</span>
<span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">coembed.sub</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Computing nearest neighbor graph</span></code></pre>
<pre><code><span class="co">## Computing SNN</span></code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">coembed.sub</span>, resolution <span class="op">=</span> <span class="fl">0.2</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">ArchR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ArchR/man/paletteDiscrete.html" class="external-link">paletteDiscrete</a></span><span class="op">(</span><span class="va">coembed.sub</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="st">"RNA_snn_res.0.2"</span><span class="op">]</span><span class="op">)</span>
    
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,
             reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"UMAP1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"UMAP2"</span><span class="op">)</span>
    
<span class="va">p</span>                </code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-14-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Marker genes are identified based on new clustering results.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">all.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">obj.coembed</span>, 
                              only.pos <span class="op">=</span> <span class="cn">TRUE</span>, 
                              min.pct <span class="op">=</span> <span class="fl">0.25</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Calculating cluster 0</span></code></pre>
<pre><code><span class="co">## Calculating cluster 1</span></code></pre>
<pre><code><span class="co">## Calculating cluster 10</span></code></pre>
<pre><code><span class="co">## Calculating cluster 11</span></code></pre>
<pre><code><span class="co">## Calculating cluster 12</span></code></pre>
<pre><code><span class="co">## Calculating cluster 13</span></code></pre>
<pre><code><span class="co">## Calculating cluster 14</span></code></pre>
<pre><code><span class="co">## Calculating cluster 15</span></code></pre>
<pre><code><span class="co">## Calculating cluster 2</span></code></pre>
<pre><code><span class="co">## Calculating cluster 3</span></code></pre>
<pre><code><span class="co">## Calculating cluster 4</span></code></pre>
<pre><code><span class="co">## Calculating cluster 5</span></code></pre>
<pre><code><span class="co">## Calculating cluster 6</span></code></pre>
<pre><code><span class="co">## Calculating cluster 7</span></code></pre>
<pre><code><span class="co">## Calculating cluster 8</span></code></pre>
<pre><code><span class="co">## Calculating cluster 9</span></code></pre>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">all.markers</span> <span class="op"><a href="https://rdrr.io/pkg/harmony/man/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/harmony/man/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, order_by <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: Scaling data with a low number of groups may produce misleading results</span></code></pre>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-15-1.png" width="1440" style="display: block; margin: auto;"></p>
<p>We can plot the snRNA-seq and snATAC-seq separately</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,
             reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, split.by <span class="op">=</span> <span class="st">"tech"</span>, 
             cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"UMAP1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"UMAP2"</span><span class="op">)</span>

<span class="va">p</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-16-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Visualize the cell proportion of each sub-population across all patients</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CellPropPlot.html">CellPropPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, 
                   group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, 
                   prop.in <span class="op">=</span> <span class="st">"patient_region_id"</span>, 
                   cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span>

<span class="va">p</span></code></pre></div>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-17-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Since we have annotated all patients into three major groups, i.e., myogenic, ischmeic, and fibrotic. we can also perform statistical test to check if any sub-popualtion are enriched in any of the above group. This can be down by the function <em>CompareCellProp</em>.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coembed.sub</span><span class="op">$</span><span class="va">patient_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">coembed.sub</span><span class="op">$</span><span class="va">patient_group</span>, 
                                    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"myogenic"</span>, <span class="st">"ischemic"</span>, <span class="st">"fibrotic"</span><span class="op">)</span><span class="op">)</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CompareCellProp.html">CompareCellProp</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">coembed.sub</span>, 
                     group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, 
                     prop.in <span class="op">=</span> <span class="st">"patient_region_id"</span>, 
                      sample.annotation <span class="op">=</span> <span class="st">"patient_group"</span>,
                    comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"myogenic"</span>, <span class="st">"ischemic"</span><span class="op">)</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ischemic"</span>, <span class="st">"fibrotic"</span><span class="op">)</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"myogenic"</span>, <span class="st">"fibrotic"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">p</span></code></pre></div>
<pre><code><span class="co">## Warning in wilcox.test.default(c(0.1147124245313, 0.121564482029598,</span>
<span class="co">## 0.120168067226891, : cannot compute exact p-value with ties</span></code></pre>
<p><img src="GRN-fibroblast_files/figure-html/unnamed-chunk-18-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="trajectory-identification">Trajectory identification<a class="anchor" aria-label="anchor" href="#trajectory-identification"></a>
</h2>
</div>
<div class="section level2">
<h2 id="tfs-and-genes-selection">TFs and genes selection<a class="anchor" aria-label="anchor" href="#tfs-and-genes-selection"></a>
</h2>
</div>
<div class="section level2">
<h2 id="gene-regulatory-network-inference-and-visualization">Gene regulatory network inference and visualization<a class="anchor" aria-label="anchor" href="#gene-regulatory-network-inference-and-visualization"></a>
</h2>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Check session information</span>
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.1 (2021-08-10)</span>
<span class="co">## Platform: x86_64-conda-linux-gnu (64-bit)</span>
<span class="co">## Running under: Rocky Linux 8.5 (Green Obsidian)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /home/rs619065/miniconda3/envs/r-4.1/lib/libopenblasp-r0.3.18.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] dplyr_1.0.9                       TFBSTools_1.32.0                 </span>
<span class="co">##  [3] JASPAR2020_0.99.10                BSgenome.Hsapiens.UCSC.hg38_1.4.4</span>
<span class="co">##  [5] BSgenome_1.62.0                   rtracklayer_1.54.0               </span>
<span class="co">##  [7] Biostrings_2.62.0                 XVector_0.34.0                   </span>
<span class="co">##  [9] GenomicRanges_1.46.1              GenomeInfoDb_1.30.1              </span>
<span class="co">## [11] IRanges_2.28.0                    S4Vectors_0.32.4                 </span>
<span class="co">## [13] BiocGenerics_0.40.0               Nebulosa_1.4.0                   </span>
<span class="co">## [15] patchwork_1.1.1                   ggplot2_3.3.6                    </span>
<span class="co">## [17] harmony_0.1.0                     Rcpp_1.0.8.3                     </span>
<span class="co">## [19] scMEGA_0.1.0                      Signac_1.6.0                     </span>
<span class="co">## [21] sp_1.4-7                          SeuratObject_4.1.0               </span>
<span class="co">## [23] Seurat_4.1.1                     </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] SnowballC_0.7.0             scattermore_0.8            </span>
<span class="co">##   [3] ggthemes_4.2.4              R.methodsS3_1.8.1          </span>
<span class="co">##   [5] ragg_1.2.2                  tidyr_1.2.0                </span>
<span class="co">##   [7] bit64_4.0.5                 knitr_1.39                 </span>
<span class="co">##   [9] R.utils_2.11.0              irlba_2.3.5                </span>
<span class="co">##  [11] DelayedArray_0.20.0         data.table_1.14.2          </span>
<span class="co">##  [13] rpart_4.1.16                KEGGREST_1.34.0            </span>
<span class="co">##  [15] RCurl_1.98-1.6              generics_0.1.2             </span>
<span class="co">##  [17] cowplot_1.1.1               RSQLite_2.2.10             </span>
<span class="co">##  [19] RANN_2.6.1                  proxy_0.4-26               </span>
<span class="co">##  [21] future_1.25.0               tzdb_0.2.0                 </span>
<span class="co">##  [23] bit_4.0.4                   spatstat.data_2.2-0        </span>
<span class="co">##  [25] httpuv_1.6.5                SummarizedExperiment_1.24.0</span>
<span class="co">##  [27] assertthat_0.2.1            DirichletMultinomial_1.36.0</span>
<span class="co">##  [29] xfun_0.31                   hms_1.1.1                  </span>
<span class="co">##  [31] jquerylib_0.1.4             evaluate_0.15              </span>
<span class="co">##  [33] promises_1.2.0.1            DEoptimR_1.0-11            </span>
<span class="co">##  [35] fansi_1.0.3                 restfulr_0.0.13            </span>
<span class="co">##  [37] caTools_1.18.2              igraph_1.3.1               </span>
<span class="co">##  [39] DBI_1.1.2                   htmlwidgets_1.5.4          </span>
<span class="co">##  [41] sparsesvd_0.2               spatstat.geom_2.4-0        </span>
<span class="co">##  [43] purrr_0.3.4                 ellipsis_0.3.2             </span>
<span class="co">##  [45] RSpectra_0.16-1             ks_1.13.5                  </span>
<span class="co">##  [47] backports_1.4.1             ggpubr_0.4.0               </span>
<span class="co">##  [49] annotate_1.72.0             deldir_1.0-6               </span>
<span class="co">##  [51] MatrixGenerics_1.6.0        vctrs_0.4.1                </span>
<span class="co">##  [53] SingleCellExperiment_1.16.0 Biobase_2.54.0             </span>
<span class="co">##  [55] TTR_0.24.3                  ROCR_1.0-11                </span>
<span class="co">##  [57] abind_1.4-5                 cachem_1.0.6               </span>
<span class="co">##  [59] RcppEigen_0.3.3.9.2         withr_2.5.0                </span>
<span class="co">##  [61] ggforce_0.3.3               robustbase_0.95-0          </span>
<span class="co">##  [63] progressr_0.10.0            vcd_1.4-9                  </span>
<span class="co">##  [65] sctransform_0.3.3           GenomicAlignments_1.30.0   </span>
<span class="co">##  [67] xts_0.12.1                  mclust_5.4.9               </span>
<span class="co">##  [69] goftest_1.2-3               cluster_2.1.2              </span>
<span class="co">##  [71] lazyeval_0.2.2              seqLogo_1.60.0             </span>
<span class="co">##  [73] laeken_0.5.2                crayon_1.5.1               </span>
<span class="co">##  [75] labeling_0.4.2              pkgconfig_2.0.3            </span>
<span class="co">##  [77] slam_0.1-50                 tweenr_1.0.2               </span>
<span class="co">##  [79] nlme_3.1-155                nnet_7.3-17                </span>
<span class="co">##  [81] rlang_1.0.2                 globals_0.15.0             </span>
<span class="co">##  [83] lifecycle_1.0.1             miniUI_0.1.1.1             </span>
<span class="co">##  [85] rprojroot_2.0.3             polyclip_1.10-0            </span>
<span class="co">##  [87] RcppHNSW_0.3.0              matrixStats_0.62.0         </span>
<span class="co">##  [89] lmtest_0.9-40               Matrix_1.4-0               </span>
<span class="co">##  [91] ggseqlogo_0.1               carData_3.0-5              </span>
<span class="co">##  [93] Rhdf5lib_1.16.0             boot_1.3-28                </span>
<span class="co">##  [95] zoo_1.8-10                  ggridges_0.5.3             </span>
<span class="co">##  [97] png_0.1-7                   viridisLite_0.4.0          </span>
<span class="co">##  [99] rjson_0.2.21                bitops_1.0-7               </span>
<span class="co">## [101] R.oo_1.24.0                 rhdf5filters_1.6.0         </span>
<span class="co">## [103] KernSmooth_2.23-20          blob_1.2.2                 </span>
<span class="co">## [105] stringr_1.4.0               parallelly_1.31.1          </span>
<span class="co">## [107] spatstat.random_2.2-0       rstatix_0.7.0              </span>
<span class="co">## [109] readr_2.1.2                 ggsignif_0.6.3             </span>
<span class="co">## [111] CNEr_1.30.0                 scales_1.2.0               </span>
<span class="co">## [113] memoise_2.0.1               magrittr_2.0.3             </span>
<span class="co">## [115] plyr_1.8.7                  hexbin_1.28.2              </span>
<span class="co">## [117] ica_1.0-2                   zlibbioc_1.40.0            </span>
<span class="co">## [119] compiler_4.1.1              BiocIO_1.4.0               </span>
<span class="co">## [121] RColorBrewer_1.1-3          pcaMethods_1.86.0          </span>
<span class="co">## [123] fitdistrplus_1.1-8          Rsamtools_2.10.0           </span>
<span class="co">## [125] cli_3.3.0                   listenv_0.8.0              </span>
<span class="co">## [127] pbapply_1.5-0               ggplot.multistats_1.0.0    </span>
<span class="co">## [129] MASS_7.3-55                 mgcv_1.8-39                </span>
<span class="co">## [131] tidyselect_1.1.2            stringi_1.7.6              </span>
<span class="co">## [133] textshaping_0.3.6           highr_0.9                  </span>
<span class="co">## [135] yaml_2.3.5                  ggrepel_0.9.1              </span>
<span class="co">## [137] grid_4.1.1                  sass_0.4.1                 </span>
<span class="co">## [139] fastmatch_1.1-3             tools_4.1.1                </span>
<span class="co">## [141] future.apply_1.9.0          parallel_4.1.1             </span>
<span class="co">## [143] TFMPvalue_0.0.8             lsa_0.73.3                 </span>
<span class="co">## [145] gridExtra_2.3               smoother_1.1               </span>
<span class="co">## [147] scatterplot3d_0.3-41        farver_2.1.0               </span>
<span class="co">## [149] Rtsne_0.16                  digest_0.6.29              </span>
<span class="co">## [151] rgeos_0.5-9                 shiny_1.7.1                </span>
<span class="co">## [153] pracma_2.3.8                qlcMatrix_0.9.7            </span>
<span class="co">## [155] broom_0.8.0                 car_3.0-13                 </span>
<span class="co">## [157] later_1.3.0                 RcppAnnoy_0.0.19           </span>
<span class="co">## [159] AnnotationDbi_1.56.2        httr_1.4.3                 </span>
<span class="co">## [161] colorspace_2.0-3            XML_3.99-0.9               </span>
<span class="co">## [163] fs_1.5.2                    tensor_1.5                 </span>
<span class="co">## [165] ranger_0.13.1               reticulate_1.24            </span>
<span class="co">## [167] splines_4.1.1               uwot_0.1.11                </span>
<span class="co">## [169] RcppRoll_0.3.0              spatstat.utils_2.3-1       </span>
<span class="co">## [171] pkgdown_2.0.3               ArchR_1.0.1                </span>
<span class="co">## [173] plotly_4.10.0               systemfonts_1.0.4          </span>
<span class="co">## [175] xtable_1.8-4                poweRlaw_0.70.6            </span>
<span class="co">## [177] jsonlite_1.8.0              destiny_3.9.1              </span>
<span class="co">## [179] R6_2.5.1                    pillar_1.7.0               </span>
<span class="co">## [181] htmltools_0.5.2             mime_0.12                  </span>
<span class="co">## [183] glue_1.6.2                  fastmap_1.1.0              </span>
<span class="co">## [185] VIM_6.1.1                   BiocParallel_1.28.3        </span>
<span class="co">## [187] class_7.3-20                codetools_0.2-18           </span>
<span class="co">## [189] mvtnorm_1.1-3               utf8_1.2.2                 </span>
<span class="co">## [191] lattice_0.20-45             bslib_0.3.1                </span>
<span class="co">## [193] spatstat.sparse_2.1-1       tibble_3.1.7               </span>
<span class="co">## [195] curl_4.3.2                  leiden_0.4.2               </span>
<span class="co">## [197] gtools_3.9.2                GO.db_3.14.0               </span>
<span class="co">## [199] limma_3.50.1                survival_3.2-13            </span>
<span class="co">## [201] rmarkdown_2.11              docopt_0.7.1               </span>
<span class="co">## [203] desc_1.4.1                  munsell_0.5.0              </span>
<span class="co">## [205] rhdf5_2.38.0                e1071_1.7-9                </span>
<span class="co">## [207] GenomeInfoDbData_1.2.7      reshape2_1.4.4             </span>
<span class="co">## [209] gtable_0.3.0                spatstat.core_2.4-2</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zhijian Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
