<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene-regulatory network of myofibroblast differentiation in myocardial infarction • scMEGA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene-regulatory network of myofibroblast differentiation in myocardial infarction">
<meta property="og:description" content="scMEGA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMEGA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/myofibroblast-GRN.html">Infer GRN for myofibroblast</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/CostaLab/scMEGA" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene-regulatory network of myofibroblast differentiation in myocardial infarction</h1>
            
            <h4 data-toc-skip class="date">Compiled: June 22, 2022</h4>
      
      
      <div class="hidden name"><code>myofibroblast-GRN.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we will use the integrated snRNA-seq and snATAC-seq data to infer a gene regulatory network to characterize myofibroblast differentiation. More details about data integration can be found <a href="https://costalab.github.io/scMEGA/articles/myofibroblast-preprocessing.html">here</a></p>
<p>More specifically, we will:</p>
<ul>
<li>Identify the trajectory for myofibroblast differentiation</li>
<li>Select the relevant TFs and genes for this process</li>
<li>Infer and visualize the gene regulatory network</li>
</ul>
<p>Next, we load all necessary packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ArchR</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/timoast/signac" class="external-link">Signac</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://costalab.github.io/scMEGA/">scMEGA</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">harmony</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/powellgenomicslab/Nebulosa" class="external-link">Nebulosa</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org" class="external-link">igraph</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We first load the integrated data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./Fibroblast/coembed.sub.rds"</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="trajectory-analysis">Trajectory analysis<a class="anchor" aria-label="anchor" href="#trajectory-analysis"></a>
</h2>
<div class="section level3">
<h3 id="dimensionality-reduction">Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h3>
<p>To infer trajectory, we will perform dimension reduction using diffusion map via the function <em>RunDiffusionMap</em>. This is based on the R package <a href="https://bioconductor.org/packages/release/bioc/html/destiny.html" class="external-link">destiny</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunDiffusionMap.html">RunDiffusionMap</a></span><span class="op">(</span><span class="va">coembed.sub</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## finding knns......done. Time: 228.98s</span>
<span class="co">## Calculating transition probabilities......done. Time: 0.26s</span>
<span class="co">## </span>
<span class="co">## performing eigen decomposition......done. Time: 1.60s</span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">ArchR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ArchR/man/paletteDiscrete.html" class="external-link">paletteDiscrete</a></span><span class="op">(</span><span class="va">coembed.sub</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="st">"RNA_snn_res.0.2"</span><span class="op">]</span><span class="op">)</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,
             reduction <span class="op">=</span> <span class="st">"dm"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"DC 1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"DC 2"</span><span class="op">)</span>

<span class="va">p</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can also plot snATAC-seq and snRNA-seq individually</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, reduction <span class="op">=</span> <span class="st">"dm"</span>, 
        group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, split.by <span class="op">=</span> <span class="st">"tech"</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-4-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="cell-pairing">Cell pairing<a class="anchor" aria-label="anchor" href="#cell-pairing"></a>
</h3>
<p>Next, we match the cells between these two modalities. In other words, for each cell in, for example, snATAC-seq, we will find a cell from snRNA-seq data so that these two cells have the similar profiles. This is only necessary when each modality was performed independently. If snRNA-seq and snATAC-seq data was generated by multi-modal protocol, e.g., 10X multiome or SHARE-seq, this step can be skipped.</p>
<p>We here use the method proposed by <a href="https://www.biorxiv.org/content/10.1101/2021.07.28.453784v1.abstract" class="external-link">Kartha, Vinay K., et al.</a> to match the cells.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df.pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PairCells.html">PairCells</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">coembed.sub</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span>,
                    pair.by <span class="op">=</span> <span class="st">"tech"</span>, ident1 <span class="op">=</span> <span class="st">"ATAC"</span>, ident2 <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Getting dimensional reduction data for pairing cells...</span></code></pre>
<pre><code><span class="co">## Pairing cells using geodesic mode...</span></code></pre>
<pre><code><span class="co">## Constructing KNN graph for computing geodesic distance ..</span></code></pre>
<pre><code><span class="co">## Computing graph-based geodesic distance ..</span></code></pre>
<pre><code><span class="co">## # KNN subgraphs detected: 1</span></code></pre>
<pre><code><span class="co">## Skipping subgraphs with either ATAC/RNA cells fewer than: 50</span></code></pre>
<pre><code><span class="co">## Pairing cells for subgraph No.1</span></code></pre>
<pre><code><span class="co">## Total ATAC cells in subgraph: 3397</span></code></pre>
<pre><code><span class="co">## Total RNA cells in subgraph: 22840</span></code></pre>
<pre><code><span class="co">## Subgraph size: 3397</span></code></pre>
<pre><code><span class="co">## Search threshold being used: 1359</span></code></pre>
<pre><code><span class="co">## Constructing KNN based on geodesic distance to reduce search pairing search space</span></code></pre>
<pre><code><span class="co">## Number of cells being paired: 3397 ATAC and 3397  RNA cells</span></code></pre>
<pre><code><span class="co">## Determing pairs through optimized bipartite matching ..</span></code></pre>
<pre><code><span class="co">## Assembling pair list ..</span></code></pre>
<pre><code><span class="co">## Finished!</span></code></pre>
<p>We can visualize the paired cells</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sel_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="va">df.pair</span><span class="op">$</span><span class="va">ATAC</span>, <span class="va">df.pair</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span>
<span class="va">coembed.sub2</span> <span class="op">&lt;-</span> <span class="va">coembed.sub</span><span class="op">[</span>, <span class="va">sel_cells</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.height <span class="op">=</span> <span class="fl">5</span>, repr.plot.width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub2</span>, reduction <span class="op">=</span> <span class="st">"dm"</span>, 
        group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, split.by <span class="op">=</span> <span class="st">"tech"</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-6-1.png" width="960" style="display: block; margin: auto;"></p>
<p>We next create a new Seurat object for there paired cells as if they are generated by single-cell multimodal protocol.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreatePairedObject.html">CreatePairedObject</a></span><span class="op">(</span>df.pair <span class="op">=</span> <span class="va">df.pair</span>, 
                               object <span class="op">=</span> <span class="va">coembed.sub2</span>,
                               use.assay1 <span class="op">=</span> <span class="st">"RNA"</span>, 
                               use.assay2 <span class="op">=</span> <span class="st">"ATAC"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Merging objects...</span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.pair</span></code></pre></div>
<pre><code><span class="co">## An object of class Seurat </span>
<span class="co">## 202776 features across 3397 samples within 2 assays </span>
<span class="co">## Active assay: RNA (28933 features, 0 variable features)</span>
<span class="co">##  1 other assay present: ATAC</span>
<span class="co">##  5 dimensional reductions calculated: pca, umap, harmony, umap_harmony, dm</span></code></pre>
<p>Finally, we infer a pseudo-time trajectory from SCARA5+ fibroblasts to myofibroblast using the approach from <a href="https://www.archrproject.com/" class="external-link">ArchR</a>. Here we modified the function to allow to take a Seurat object as input</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTrajectory.html">AddTrajectory</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj.pair</span>, 
                          trajectory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>,
                          group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, 
                          reduction <span class="op">=</span> <span class="st">"dm"</span>,
                          dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, 
                          use.all <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
                          
<span class="co"># we only plot the cells that are in this trajectory</span>
<span class="va">obj</span> <span class="op">&lt;-</span> <span class="va">obj.pair</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">obj.pair</span><span class="op">$</span><span class="va">Trajectory</span><span class="op">)</span><span class="op">]</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TrajectoryPlot.html">TrajectoryPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, 
                    reduction <span class="op">=</span> <span class="st">"dm"</span>,
                    continuousSet <span class="op">=</span> <span class="st">"blueYellow"</span>,
                    size <span class="op">=</span> <span class="fl">1</span>,
                   addArrow <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">p</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-8-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="tf-and-gene-selection">TF and gene selection<a class="anchor" aria-label="anchor" href="#tf-and-gene-selection"></a>
</h2>
<p>We next select candidate TFs and genes for building a meaningful gene regulatory network.</p>
<div class="section level3">
<h3 id="select-tfs">Select TFs<a class="anchor" aria-label="anchor" href="#select-tfs"></a>
</h3>
<p>To identify potential regulator (i.e., TFs), we first estimate an acitivty score for each TF in each cell. This is done by first performing motif matching and then computing deviation scores using <a href="https://greenleaflab.github.io/chromVAR/index.html" class="external-link">chromVAR</a>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get a list of motif position frequency matrices from the JASPAR database</span>
<span class="va">pfm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html" class="external-link">getMatrixSet</a></span><span class="op">(</span>
  x <span class="op">=</span> <span class="va">JASPAR2020</span>,
  opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>collection <span class="op">=</span> <span class="st">"CORE"</span>, tax_group <span class="op">=</span> <span class="st">'vertebrates'</span>, all_versions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># add motif information</span>
<span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/signac/reference/AddMotifs.html" class="external-link">AddMotifs</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">obj</span>,
  genome <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span>,
  pfm <span class="op">=</span> <span class="va">pfm</span>,
    assay <span class="op">=</span> <span class="st">"ATAC"</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Building motif matrix</span></code></pre>
<pre><code><span class="co">## Finding motif positions</span></code></pre>
<pre><code><span class="co">## Creating Motif object</span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/signac/reference/RunChromVAR.html" class="external-link">RunChromVAR</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">obj</span>,
  genome <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span>,
    assay <span class="op">=</span> <span class="st">"ATAC"</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Computing GC bias per region</span></code></pre>
<pre><code><span class="co">## Selecting background regions</span></code></pre>
<pre><code><span class="co">## Computing deviations from background</span></code></pre>
<pre><code><span class="co">## Constructing chromVAR assay</span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SelectTFs.html">SelectTFs</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, return.heatmap <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Find 456 shared features!</span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df.cor</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">tfs</span>
<span class="va">ht</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">heatmap</span></code></pre></div>
<p>We can visualize the TF activity dynamic along the trajectory</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">draw</span><span class="op">(</span><span class="va">ht</span><span class="op">)</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-10-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="select-genes">Select genes<a class="anchor" aria-label="anchor" href="#select-genes"></a>
</h3>
<p>We will select relevant genes by first linking genes to peaks based on the corrleation between gene expression from the snRNA-seq data and peak accessibility from the snATAC-seq data along the inferred trajectory. This means that we only consider a gene to be a potential target if it can be assocaited to at least one peak.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SelectGenes.html">SelectGenes</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>,
                  labelTop1 <span class="op">=</span> <span class="fl">0</span>,
                  labelTop2 <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Creating Trajectory Group Matrix..</span></code></pre>
<pre><code><span class="co">## Smoothing...</span></code></pre>
<pre><code><span class="co">## Creating Trajectory Group Matrix..</span></code></pre>
<pre><code><span class="co">## Smoothing...</span></code></pre>
<pre><code><span class="co">## Linking cis-regulatory elements to genes...</span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df.p2g</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">p2g</span>
<span class="va">ht</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">heatmap</span>

<span class="fu">draw</span><span class="op">(</span><span class="va">ht</span><span class="op">)</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-11-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="gene-regulatory-network-inference">Gene regulatory network inference<a class="anchor" aria-label="anchor" href="#gene-regulatory-network-inference"></a>
</h2>
<p>We here will try to predict a gene regulatory network based on the correlation of the TF binding activity as estimated from snATAC-seq and gene expression as measured by snRNA-seq along the trajectory.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tf.gene.cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetTFGeneCorrelation.html">GetTFGeneCorrelation</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, 
                                    tf.use <span class="op">=</span> <span class="va">df.cor</span><span class="op">$</span><span class="va">tfs</span>, 
                                    gene.use <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df.p2g</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>,
                                    tf.assay <span class="op">=</span> <span class="st">"chromvar"</span>, 
                                    gene.assay <span class="op">=</span> <span class="st">"RNA"</span>,
                                    trajectory.name <span class="op">=</span> <span class="st">"Trajectory"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Creating Trajectory Group Matrix..</span></code></pre>
<pre><code><span class="co">## Some values are below 0, this could be the Motif activity matrix in which scaleTo should be set = NULL.</span>
<span class="co">## Continuing without depth normalization!</span></code></pre>
<pre><code><span class="co">## Smoothing...</span></code></pre>
<pre><code><span class="co">## Creating Trajectory Group Matrix..</span></code></pre>
<pre><code><span class="co">## Smoothing...</span></code></pre>
<p>We can then visualize this correlation matrix by heatmap. Also, we can cluster the genes and TFs to identify different regulatory modules for the predefined sub-populations.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ht</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRNHeatmap.html">GRNHeatmap</a></span><span class="op">(</span><span class="va">tf.gene.cor</span>, 
                 tf.timepoint <span class="op">=</span> <span class="va">df.cor</span><span class="op">$</span><span class="va">time_point</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `use_raster` is automatically set to TRUE for a matrix with more than</span>
<span class="co">## 2000 rows. You can control `use_raster` argument by explicitly setting</span>
<span class="co">## TRUE/FALSE to it.</span>
<span class="co">## </span>
<span class="co">## Set `ht_opt$message = FALSE` to turn off this message.</span></code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ht</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-13-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>To associate genes to TFs, we will use the peak-to-gene links and TF binding sites information. Specifically, if a gene is regulated by a peak and this peak is bound by a TF, then we consider this gene to be a target of this TF.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">motif.matching</span> <span class="op">&lt;-</span> <span class="va">obj</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ATAC</span><span class="op">@</span><span class="va">motifs</span><span class="op">@</span><span class="va">data</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">motif.matching</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">obj</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ATAC</span><span class="op">@</span><span class="va">motifs</span><span class="op">@</span><span class="va">motif.names</span>
<span class="va">motif.matching</span> <span class="op">&lt;-</span>
    <span class="va">motif.matching</span><span class="op">[</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df.p2g</span><span class="op">$</span><span class="va">peak</span><span class="op">)</span>, <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">tf.gene.cor</span><span class="op">$</span><span class="va">tf</span><span class="op">)</span><span class="op">]</span>


<span class="va">df.grn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetGRN.html">GetGRN</a></span><span class="op">(</span>motif.matching <span class="op">=</span> <span class="va">motif.matching</span>, 
                 df.cor <span class="op">=</span> <span class="va">tf.gene.cor</span>, 
                 df.p2g <span class="op">=</span> <span class="va">df.p2g</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Filtering network by peak-to-gene links...</span></code></pre>
<pre><code><span class="co">## Filtering network by TF binding site prediction...</span></code></pre>
<p>Next, we can visualize our network as the last step of this analysis</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># define colors for nodes representing TFs (i.e., regulators)</span>
<span class="va">df.cor</span> <span class="op">&lt;-</span> <span class="va">df.cor</span><span class="op">[</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">order</a></span><span class="op">(</span><span class="va">df.cor</span><span class="op">$</span><span class="va">time_point</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">tfs.timepoint</span> <span class="op">&lt;-</span> <span class="va">df.cor</span><span class="op">$</span><span class="va">time_point</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tfs.timepoint</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">df.cor</span><span class="op">$</span><span class="va">tfs</span>

<span class="co"># plot the graph, here we can highlight some genes</span>
<span class="va">df.grn2</span> <span class="op">&lt;-</span> <span class="va">df.grn</span> <span class="op"><a href="https://rdrr.io/pkg/harmony/man/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://satijalab.org/signac/reference/subset.Motif.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">correlation</span> <span class="op">&gt;</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/harmony/man/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="va">tf</span>, <span class="va">gene</span>, <span class="va">correlation</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/harmony/man/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>weights <span class="op">=</span> <span class="va">correlation</span><span class="op">)</span>
    
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRNPlot.html">GRNPlot</a></span><span class="op">(</span><span class="va">df.grn2</span>, 
             tfs.timepoint <span class="op">=</span> <span class="va">tfs.timepoint</span>,
             show.tf.labels <span class="op">=</span> <span class="cn">TRUE</span>,
             seed <span class="op">=</span> <span class="fl">42</span>, 
             plot.importance <span class="op">=</span> <span class="cn">FALSE</span>,
            min.importance <span class="op">=</span> <span class="fl">2</span>,
            remove.isolated <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.height <span class="op">=</span> <span class="fl">20</span>, repr.plot.width <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: Using alpha for a discrete variable is not advised.</span></code></pre>
<p><img src="myofibroblast-GRN_files/figure-html/viz_network-1.png" width="1440" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="grn-visualization">GRN visualization<a class="anchor" aria-label="anchor" href="#grn-visualization"></a>
</h2>
<div class="section level3">
<h3 id="grn-along-the-trajectory">GRN along the trajectory<a class="anchor" aria-label="anchor" href="#grn-along-the-trajectory"></a>
</h3>
</div>
<div class="section level3">
<h3 id="grn-in-space">GRN in space<a class="anchor" aria-label="anchor" href="#grn-in-space"></a>
</h3>
<p>As an additional step, if the spatial transcriptome (ST) data (e.g., generated by using 10X Visium assay) are available, we can then visualize the target gene expression for some interesting TFs to understand the regulon activity in spatial space. First let’s download the ST data</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> VisiumSpatial</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-P</span> VisiumSpatial https://costalab.ukaachen.de/open_data/scMEGA/VisiumSpatial/AKK004_157772.rds</span></code></pre></div>
<p>Of note, each spot in the 10X Visium assay contains multiple cells, meaning that we usually cannot identify cell types by clustering the data. To address this issue, people developed various computation tools to estimate cell-type-specific abundance in each spot. Here we have perform such analysis using <a href="https://www.nature.com/articles/s41587-021-01139-4" class="external-link">cell2location</a> based on our snRNA-seq data. Let’s first visualize the fibroblast proportion.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./VisiumSpatial/AKK004_157772.rds"</span><span class="op">)</span>

<span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">obj.spatial</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"c2l_props"</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialFeaturePlot</a></span><span class="op">(</span><span class="va">obj.spatial</span>, features <span class="op">=</span> <span class="st">"Fib"</span>, min.cutoff <span class="op">=</span> <span class="st">"q5"</span>,
                       max.cutoff <span class="op">=</span> <span class="st">"q95"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"D"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Scale for 'fill' is already present. Adding another scale for 'fill', which</span>
<span class="co">## will replace the existing scale.</span></code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-16-1.png" width="480" style="display: block; margin: auto;"></p>
<p>We next can check the gene expression in space. Here, we use the above TFs.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">obj.spatial</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT"</span>

<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialFeaturePlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj.spatial</span>, features <span class="op">=</span> <span class="st">"NR3C2"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span>
<span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"C"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Scale for 'fill' is already present. Adding another scale for 'fill', which</span>
<span class="co">## will replace the existing scale.</span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialFeaturePlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj.spatial</span>, features <span class="op">=</span> <span class="st">"RUNX1"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span>
<span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"C"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Scale for 'fill' is already present. Adding another scale for 'fill', which</span>
<span class="co">## will replace the existing scale.</span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-17-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Then, we visualize the regulon activity of NR3C2 and RUNX1 by using the GRN that we predicted. The major idea is that if the target genes of a TF are up-regulated, then this TF should have high activity. This principle has been used to develop computation methods to estimate TF activity from gene expression data, such as <a href="https://genome.cshlp.org/content/29/8/1363.short" class="external-link">DoRothEA</a>.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">obj.spatial</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT"</span>

<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRNSpatialPlot.html">GRNSpatialPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj.spatial</span>, assay <span class="op">=</span> <span class="st">"SCT"</span>, df.grn <span class="op">=</span> <span class="va">df.grn2</span>,
                   tf.use <span class="op">=</span> <span class="st">"NR3C2"</span>, vis.option <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## Warning: The following features are not present in the object: CATSPERB, GALR1,</span>
<span class="co">## LMX1A, TEX26-AS1, UGDH-AS1, ZBTB20-AS5, not searching for symbol synonyms</span></code></pre>
<pre><code><span class="co">## Scale for 'fill' is already present. Adding another scale for 'fill', which</span>
<span class="co">## will replace the existing scale.</span></code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRNSpatialPlot.html">GRNSpatialPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj.spatial</span>, assay <span class="op">=</span> <span class="st">"SCT"</span>, df.grn <span class="op">=</span> <span class="va">df.grn2</span>,
                   tf.use <span class="op">=</span> <span class="st">"RUNX1"</span>, vis.option <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: The following features are not present in the object: CNTNAP2, DEC1,</span>
<span class="co">## HMGA2, MIR181A1HG, RFX8, TPRG1, VEPH1, not searching for symbol synonyms</span></code></pre>
<pre><code><span class="co">## Scale for 'fill' is already present. Adding another scale for 'fill', which</span>
<span class="co">## will replace the existing scale.</span></code></pre>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-18-1.png" width="960" style="display: block; margin: auto;"></p>
<p>The above plots demonstrate a clear spatial-distributed pattern for the regulon activity of NR3C2 and RUNX1.</p>
<p>Save data</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">obj</span>, <span class="st">"./Fibroblast/coembed.trajectory.rds"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Check session information</span>
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.1 (2021-08-10)</span>
<span class="co">## Platform: x86_64-conda-linux-gnu (64-bit)</span>
<span class="co">## Running under: Rocky Linux 8.5 (Green Obsidian)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /home/rs619065/miniconda3/envs/r-4.1/lib/libopenblasp-r0.3.18.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">## [8] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] circlize_0.4.15                   ComplexHeatmap_2.10.0            </span>
<span class="co">##  [3] ggraph_2.0.5                      igraph_1.3.2                     </span>
<span class="co">##  [5] dplyr_1.0.9                       TFBSTools_1.32.0                 </span>
<span class="co">##  [7] JASPAR2020_0.99.10                BSgenome.Hsapiens.UCSC.hg38_1.4.4</span>
<span class="co">##  [9] BSgenome_1.62.0                   rtracklayer_1.54.0               </span>
<span class="co">## [11] Biostrings_2.62.0                 XVector_0.34.0                   </span>
<span class="co">## [13] Nebulosa_1.4.0                    patchwork_1.1.1                  </span>
<span class="co">## [15] harmony_0.1.0                     Rcpp_1.0.8.3                     </span>
<span class="co">## [17] scMEGA_0.1.0                      Signac_1.6.0                     </span>
<span class="co">## [19] sp_1.5-0                          SeuratObject_4.1.0               </span>
<span class="co">## [21] Seurat_4.1.1                      ArchR_1.0.1                      </span>
<span class="co">## [23] magrittr_2.0.3                    rhdf5_2.38.1                     </span>
<span class="co">## [25] Matrix_1.4-1                      data.table_1.14.2                </span>
<span class="co">## [27] SummarizedExperiment_1.24.0       Biobase_2.54.0                   </span>
<span class="co">## [29] GenomicRanges_1.46.1              GenomeInfoDb_1.30.1              </span>
<span class="co">## [31] IRanges_2.28.0                    S4Vectors_0.32.4                 </span>
<span class="co">## [33] BiocGenerics_0.40.0               MatrixGenerics_1.6.0             </span>
<span class="co">## [35] matrixStats_0.62.0                ggplot2_3.3.6                    </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] vcd_1.4-9                   ica_1.0-2                  </span>
<span class="co">##   [3] RcppRoll_0.3.0              class_7.3-20               </span>
<span class="co">##   [5] Rsamtools_2.10.0            foreach_1.5.2              </span>
<span class="co">##   [7] lmtest_0.9-40               rprojroot_2.0.3            </span>
<span class="co">##   [9] crayon_1.5.1                laeken_0.5.2               </span>
<span class="co">##  [11] spatstat.core_2.4-4         MASS_7.3-57                </span>
<span class="co">##  [13] rhdf5filters_1.6.0          nlme_3.1-157               </span>
<span class="co">##  [15] qlcMatrix_0.9.7             rlang_1.0.2                </span>
<span class="co">##  [17] ROCR_1.0-11                 irlba_2.3.5                </span>
<span class="co">##  [19] smoother_1.1                BiocParallel_1.28.3        </span>
<span class="co">##  [21] rjson_0.2.21                CNEr_1.30.0                </span>
<span class="co">##  [23] bit64_4.0.5                 glue_1.6.2                 </span>
<span class="co">##  [25] poweRlaw_0.70.6             sctransform_0.3.3          </span>
<span class="co">##  [27] parallel_4.1.1              chromVAR_1.16.0            </span>
<span class="co">##  [29] spatstat.sparse_2.1-1       AnnotationDbi_1.56.2       </span>
<span class="co">##  [31] spatstat.geom_2.4-0         tidyselect_1.1.2           </span>
<span class="co">##  [33] textshape_1.7.3             fitdistrplus_1.1-8         </span>
<span class="co">##  [35] XML_3.99-0.9                tidyr_1.2.0                </span>
<span class="co">##  [37] zoo_1.8-10                  GenomicAlignments_1.30.0   </span>
<span class="co">##  [39] xtable_1.8-4                RcppHNSW_0.3.0             </span>
<span class="co">##  [41] evaluate_0.15               cli_3.3.0                  </span>
<span class="co">##  [43] zlibbioc_1.40.0             miniUI_0.1.1.1             </span>
<span class="co">##  [45] bslib_0.3.1                 rpart_4.1.16               </span>
<span class="co">##  [47] fastmatch_1.1-3             RcppEigen_0.3.3.9.2        </span>
<span class="co">##  [49] shiny_1.7.1                 xfun_0.31                  </span>
<span class="co">##  [51] clue_0.3-60                 cluster_2.1.3              </span>
<span class="co">##  [53] caTools_1.18.2              tidygraph_1.2.1            </span>
<span class="co">##  [55] pcaMethods_1.86.0           KEGGREST_1.34.0            </span>
<span class="co">##  [57] tibble_3.1.7                ggrepel_0.9.1              </span>
<span class="co">##  [59] listenv_0.8.0               TFMPvalue_0.0.8            </span>
<span class="co">##  [61] png_0.1-7                   future_1.26.1              </span>
<span class="co">##  [63] withr_2.5.0                 lsa_0.73.3                 </span>
<span class="co">##  [65] bitops_1.0-7                slam_0.1-50                </span>
<span class="co">##  [67] ggforce_0.3.3               ranger_0.13.1              </span>
<span class="co">##  [69] plyr_1.8.7                  sparsesvd_0.2              </span>
<span class="co">##  [71] e1071_1.7-9                 pracma_2.3.8               </span>
<span class="co">##  [73] pillar_1.7.0                GlobalOptions_0.1.2        </span>
<span class="co">##  [75] cachem_1.0.6                fs_1.5.2                   </span>
<span class="co">##  [77] scatterplot3d_0.3-41        TTR_0.24.3                 </span>
<span class="co">##  [79] GetoptLong_1.0.5            xts_0.12.1                 </span>
<span class="co">##  [81] vctrs_0.4.1                 ellipsis_0.3.2             </span>
<span class="co">##  [83] generics_0.1.2              tools_4.1.1                </span>
<span class="co">##  [85] munsell_0.5.0               tweenr_1.0.2               </span>
<span class="co">##  [87] proxy_0.4-26                DelayedArray_0.20.0        </span>
<span class="co">##  [89] fastmap_1.1.0               compiler_4.1.1             </span>
<span class="co">##  [91] abind_1.4-5                 httpuv_1.6.5               </span>
<span class="co">##  [93] plotly_4.10.0               rgeos_0.5-9                </span>
<span class="co">##  [95] GenomeInfoDbData_1.2.7      gridExtra_2.3              </span>
<span class="co">##  [97] lattice_0.20-45             deldir_1.0-6               </span>
<span class="co">##  [99] utf8_1.2.2                  later_1.3.0                </span>
<span class="co">## [101] jsonlite_1.8.0              ggplot.multistats_1.0.0    </span>
<span class="co">## [103] scales_1.2.0                docopt_0.7.1               </span>
<span class="co">## [105] pbapply_1.5-0               carData_3.0-5              </span>
<span class="co">## [107] lazyeval_0.2.2              nabor_0.5.0                </span>
<span class="co">## [109] promises_1.2.0.1            car_3.1-0                  </span>
<span class="co">## [111] doParallel_1.0.17           R.utils_2.11.0             </span>
<span class="co">## [113] goftest_1.2-3               spatstat.utils_2.3-1       </span>
<span class="co">## [115] reticulate_1.25             rmarkdown_2.14             </span>
<span class="co">## [117] pkgdown_2.0.3               rlemon_0.2.0               </span>
<span class="co">## [119] cowplot_1.1.1               textshaping_0.3.6          </span>
<span class="co">## [121] Rtsne_0.16                  uwot_0.1.11                </span>
<span class="co">## [123] survival_3.3-1              yaml_2.3.5                 </span>
<span class="co">## [125] systemfonts_1.0.4           htmltools_0.5.2            </span>
<span class="co">## [127] memoise_2.0.1               BiocIO_1.4.0               </span>
<span class="co">## [129] graphlayouts_0.8.0          destiny_3.9.1              </span>
<span class="co">## [131] viridisLite_0.4.0           digest_0.6.29              </span>
<span class="co">## [133] assertthat_0.2.1            mime_0.12                  </span>
<span class="co">## [135] RSQLite_2.2.14              future.apply_1.9.0         </span>
<span class="co">## [137] blob_1.2.3                  R.oo_1.24.0                </span>
<span class="co">## [139] ragg_1.2.2                  splines_4.1.1              </span>
<span class="co">## [141] labeling_0.4.2              Rhdf5lib_1.16.0            </span>
<span class="co">## [143] RCurl_1.98-1.6              ks_1.13.5                  </span>
<span class="co">## [145] hms_1.1.1                   colorspace_2.0-3           </span>
<span class="co">## [147] shape_1.4.6                 nnet_7.3-17                </span>
<span class="co">## [149] sass_0.4.1                  mclust_5.4.10              </span>
<span class="co">## [151] RANN_2.6.1                  mvtnorm_1.1-3              </span>
<span class="co">## [153] ggseqlogo_0.1               fansi_1.0.3                </span>
<span class="co">## [155] tzdb_0.3.0                  VIM_6.1.1                  </span>
<span class="co">## [157] parallelly_1.32.0           SnowballC_0.7.0            </span>
<span class="co">## [159] R6_2.5.1                    ggridges_0.5.3             </span>
<span class="co">## [161] lifecycle_1.0.1             curl_4.3.2                 </span>
<span class="co">## [163] leiden_0.4.2                motifmatchr_1.16.0         </span>
<span class="co">## [165] jquerylib_0.1.4             robustbase_0.95-0          </span>
<span class="co">## [167] desc_1.4.1                  RcppAnnoy_0.0.19           </span>
<span class="co">## [169] RColorBrewer_1.1-3          iterators_1.0.14           </span>
<span class="co">## [171] stringr_1.4.0               htmlwidgets_1.5.4          </span>
<span class="co">## [173] polyclip_1.10-0             purrr_0.3.4                </span>
<span class="co">## [175] seqLogo_1.60.0              mgcv_1.8-40                </span>
<span class="co">## [177] globals_0.15.0              spatstat.random_2.2-0      </span>
<span class="co">## [179] progressr_0.10.1            codetools_0.2-18           </span>
<span class="co">## [181] GO.db_3.14.0                gtools_3.9.2.2             </span>
<span class="co">## [183] SingleCellExperiment_1.16.0 RSpectra_0.16-1            </span>
<span class="co">## [185] R.methodsS3_1.8.1           gtable_0.3.0               </span>
<span class="co">## [187] DBI_1.1.2                   tensor_1.5                 </span>
<span class="co">## [189] httr_1.4.3                  highr_0.9                  </span>
<span class="co">## [191] KernSmooth_2.23-20          stringi_1.7.6              </span>
<span class="co">## [193] reshape2_1.4.4              farver_2.1.0               </span>
<span class="co">## [195] annotate_1.72.0             viridis_0.6.2              </span>
<span class="co">## [197] ggthemes_4.2.4              hexbin_1.28.2              </span>
<span class="co">## [199] optmatch_0.10.3             magick_2.7.3               </span>
<span class="co">## [201] DT_0.23                     boot_1.3-28                </span>
<span class="co">## [203] restfulr_0.0.13             readr_2.1.2                </span>
<span class="co">## [205] scattermore_0.8             DEoptimR_1.0-11            </span>
<span class="co">## [207] bit_4.0.4                   spatstat.data_2.2-0        </span>
<span class="co">## [209] pkgconfig_2.0.3             DirichletMultinomial_1.36.0</span>
<span class="co">## [211] knn.covertree_1.0           knitr_1.39</span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zhijian Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
