<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene-regulatory network of myofibroblast differentiation in myocardial infarction • scMEGA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene-regulatory network of myofibroblast differentiation in myocardial infarction">
<meta property="og:description" content="scMEGA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMEGA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/myofibroblast-GRN.html">Infer GRN for myofibroblast</a>
    </li>
    <li>
      <a href="../articles/cardiomyocyte-GRN.html">Infer GRN for cardiomyocytes</a>
    </li>
    <li>
      <a href="../articles/pbmc_10x_multiome.html">Infer GRN for PBMC cells</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/CostaLab/scMEGA" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene-regulatory network of myofibroblast
differentiation in myocardial infarction</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 01, 2024</h4>
      
      
      <div class="hidden name"><code>myofibroblast-GRN.Rmd</code></div>

    </div>

    
    
<p>During this tutorial, we will integrate the snRNA-seq and snATAC-seq
data generated from the human heart samples after myocardial infarction.
The integrated data will be used as input for inferring gene regulatory
network.</p>
<p>We first download the required data. In this case, we need two Seurat
objects with each one corresponding to snRNA-seq and snATAC-seq
respectively. The snRNA-seq object includes gene expression data of all
fibroblasts and the snATAC-seq includes all chromatin accessibility
profiles. Additionally, we also need a gene activity matrix for data
integration. This matrix was estimated from the snATAC-seq data by using
the <a href="https://www.archrproject.com/" class="external-link">ArchR</a> package. The
script of cleaning the data and preparing these objects is found <a href="https://costalab.ukaachen.de/open_data/scMEGA/Fibroblast/01_prepare_data.html" class="external-link">here</a>.</p>
<p>Run the following commands to download the data:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">'./Fibroblast'</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">'./Fibroblast'</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">'https://costalab.ukaachen.de/open_data/scMEGA/Fibroblast/snRNA.rds'</span>, </span>
<span>              destfile <span class="op">=</span> <span class="st">'./Fibroblast/snRNA.rds'</span>, </span>
<span>              method <span class="op">=</span> <span class="st">'wget'</span>, extra <span class="op">=</span> <span class="st">'--no-check-certificate'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">'https://costalab.ukaachen.de/open_data/scMEGA/Fibroblast/snATAC.rds'</span>, </span>
<span>              destfile <span class="op">=</span> <span class="st">'./Fibroblast/snATAC.rds'</span>, </span>
<span>              method <span class="op">=</span> <span class="st">'wget'</span>, extra <span class="op">=</span> <span class="st">'--no-check-certificate'</span><span class="op">)</span></span>
<span>              </span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">'https://costalab.ukaachen.de/open_data/scMEGA/Fibroblast/gene.activity.rds'</span>, </span>
<span>              destfile <span class="op">=</span> <span class="st">'./Fibroblast/gene.activity.rds'</span>, </span>
<span>              method <span class="op">=</span> <span class="st">'wget'</span>, extra <span class="op">=</span> <span class="st">'--no-check-certificate'</span><span class="op">)</span></span></code></pre></div>
<p>Next, we load all necessary packages:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ArchR</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://costalab.github.io/scMEGA/">scMEGA</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">harmony</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/powellgenomicslab/Nebulosa" class="external-link">Nebulosa</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize <span class="op">=</span> <span class="fl">1e9</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="data-integration">Data integration<a class="anchor" aria-label="anchor" href="#data-integration"></a>
</h2>
<p>Let’s load the data into memory and see how they look like</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./Fibroblast/snRNA.rds"</span><span class="op">)</span></span>
<span><span class="va">obj.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./Fibroblast/snATAC.rds"</span><span class="op">)</span></span>
<span><span class="va">gene.activity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./Fibroblast/gene.activity.rds"</span><span class="op">)</span></span></code></pre></div>
<p>We need to convert the assays to an Assay5 for Seuratv5</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.rna</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">obj.rna</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Assay5"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Assay RNA changing from Assay to Assay5</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.atac</span><span class="op">[[</span><span class="st">"ATAC"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">obj.atac</span><span class="op">[[</span><span class="st">"ATAC"</span><span class="op">]</span><span class="op">]</span>, <span class="st">"Assay5"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Assay ATAC changing from ChromatinAssay to Assay5</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.rna</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 28933 features across 45515 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (28933 features, 2000 variable features)</span></span>
<span><span class="co">##  3 layers present: data, counts, scale.data</span></span>
<span><span class="co">##  4 dimensional reductions calculated: pca, harmony, umap, umap_harmony</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.atac</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 173843 features across 6481 samples within 1 assay </span></span>
<span><span class="co">## Active assay: ATAC (173843 features, 0 variable features)</span></span>
<span><span class="co">##  2 layers present: counts, data</span></span>
<span><span class="co">##  3 dimensional reductions calculated: harmony, umap, umap_harmony</span></span></code></pre>
<p>We can observe that there are 45,515 and 6,481 cells in our snRNA-seq
and snATAC-seq datasets. Let’s now visualize the data as colored by
patients. Note that here we used the UMAP embedding generated from
batch-corrected low-dimensional space so that no batch effects are
observed from the 2D visualization.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.rna</span>, pt.size <span class="op">=</span> <span class="fl">1</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"snRNA-seq"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.atac</span>, pt.size <span class="op">=</span> <span class="fl">1</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"snATAC-seq"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-1-1.png" width="1152" style="display: block; margin: auto;"></p>
<div class="section level3">
<h3 id="co-embedding">Co-embedding<a class="anchor" aria-label="anchor" href="#co-embedding"></a>
</h3>
<p>First, we need to project the data into a common low-dimensional
space. This is done by using the CCA method from <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>. To this end, we have
wrapped several functions from Seurat into a single function
<em>CoembedData</em>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoembedData.html">CoembedData</a></span><span class="op">(</span></span>
<span>  <span class="va">obj.rna</span>,</span>
<span>  <span class="va">obj.atac</span>, </span>
<span>  <span class="va">gene.activity</span>, </span>
<span>  weight.reduction <span class="op">=</span> <span class="st">"harmony"</span>, </span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Find 19026 common genes between ATAC and RNA</span></span></code></pre>
<pre><code><span><span class="co">## Subset ATAC and RNA data</span></span></code></pre>
<pre><code><span><span class="co">## Normalize gene activity score for ATAC data</span></span></code></pre>
<pre><code><span><span class="co">## Performing data integration using Seurat...</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Different features in new layer data than already exists for</span></span>
<span><span class="co">## scale.data</span></span></code></pre>
<pre><code><span><span class="co">## Finding integration vectors</span></span></code></pre>
<pre><code><span><span class="co">## Finding integration vector weights</span></span></code></pre>
<pre><code><span><span class="co">## Transfering 19026 features onto reference data</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Layer counts isn't present in the assay object; returning NULL</span></span></code></pre>
<pre><code><span><span class="co">## Coemebdding the data...</span></span></code></pre>
<pre><code><span><span class="co">## Centering data matrix</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Different features in new layer data than already exists for</span></span>
<span><span class="co">## scale.data</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Different cells in new layer data than already exists for scale.data</span></span></code></pre>
<pre><code><span><span class="co">## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">## This message will be shown once per session</span></span></code></pre>
<p>We next visualize the snRNA-seq and snATAC-seq in this shared UMAP
space. The cells are colored by patients or modalities.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"patient"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"tech"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-2-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>The batch effects between patients, regions and modalities are quite
clear. So next we use <a href="https://www.nature.com/articles/s41592-019-0619-0" class="external-link">Harmony</a> to
perform batch correction and generate a new UMAP embedding.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/harmony/man/RunHarmony.html" class="external-link">RunHarmony</a></span><span class="op">(</span><span class="va">obj.coembed</span>, </span>
<span>                          group.by.vars <span class="op">=</span> <span class="st">"patient"</span>,</span>
<span>                          reduction.use <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                          dims.use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>                          project.dim <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          plot_convergence <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Transposing data matrix</span></span></code></pre>
<pre><code><span><span class="co">## Initializing state using k-means centroids initialization</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Quick-TRANSfer stage steps exceeded maximum (= 2599800)</span></span></code></pre>
<pre><code><span><span class="co">## Warning: did not converge in 25 iterations</span></span>
<span><span class="co">## Warning: did not converge in 25 iterations</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Quick-TRANSfer stage steps exceeded maximum (= 2599800)</span></span>
<span><span class="co">## Warning: Quick-TRANSfer stage steps exceeded maximum (= 2599800)</span></span>
<span><span class="co">## Warning: Quick-TRANSfer stage steps exceeded maximum (= 2599800)</span></span></code></pre>
<pre><code><span><span class="co">## Harmony 1/10</span></span></code></pre>
<pre><code><span><span class="co">## Harmony 2/10</span></span></code></pre>
<pre><code><span><span class="co">## Harmony converged after 2 iterations</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">obj.coembed</span>,</span>
<span>  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>  reduction <span class="op">=</span> <span class="st">'harmony'</span>,</span>
<span>  reduction.name <span class="op">=</span> <span class="st">"umap_harmony"</span>,</span>
<span>  reduction.key <span class="op">=</span> <span class="st">'umapharmony_'</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can plot the data again</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"patient"</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"tech"</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-3-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>From the new UMAP embedding, we can observe that after
batch-correction, cells from different patients, regions, and modalities
are well mixed.</p>
<p>Based on our previous works of myofibroblast differentiation in <a href="https://www.nature.com/articles/s41586-020-2941-1" class="external-link">human</a> and
<a href="https://www.nature.com/articles/s41467-021-26530-2" class="external-link">mouse</a>
kidney, we already known some relevant genes for this biological
process. For example, SCARA5 is a marker for myofibroblast progenitor,
and COL1A1, POSTN, and FN1 are highly expressed in myofibroblast.
Therefore we can visualize the expression of these genes to check if we
can also identify similar process in human heart. Note that to make the
visualization clear, here we used the package <a href="https://github.com/powellgenomicslab/Nebulosa" class="external-link">Nebulosa</a> to
plot the data.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/Nebulosa/man/plot_density.html" class="external-link">plot_density</a></span><span class="op">(</span><span class="va">obj.coembed</span>,</span>
<span>               features <span class="op">=</span> <span class="st">"SCARA5"</span>,</span>
<span>               reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>,</span>
<span>               pal <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/Nebulosa/man/plot_density.html" class="external-link">plot_density</a></span><span class="op">(</span><span class="va">obj.coembed</span>,</span>
<span>               features <span class="op">=</span> <span class="st">"COL1A1"</span>,</span>
<span>               reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>,</span>
<span>               pal <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/Nebulosa/man/plot_density.html" class="external-link">plot_density</a></span><span class="op">(</span><span class="va">obj.coembed</span>,</span>
<span>               features <span class="op">=</span> <span class="st">"POSTN"</span>,</span>
<span>               reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>,</span>
<span>               pal <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/Nebulosa/man/plot_density.html" class="external-link">plot_density</a></span><span class="op">(</span><span class="va">obj.coembed</span>,</span>
<span>               features <span class="op">=</span> <span class="st">"FN1"</span>,</span>
<span>               reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>,</span>
<span>               pal <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">p3</span> <span class="op">+</span> <span class="va">p4</span><span class="op">)</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-4-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>From the visualization, we can observe that some cells highly express
SCARA5 which could be the progenitors of myofibroblasts. On the other
hand, some cells highly express COL1A1, POSTN, and FN1 and they could be
terminally differentiated myofibroblasts.</p>
</div>
<div class="section level3">
<h3 id="sub-clustering">Sub-clustering<a class="anchor" aria-label="anchor" href="#sub-clustering"></a>
</h3>
<p>We next perform sub-clustering to identify different populations in
our multi-omic fibroblast data. To further control the data quality,
here we will use a two-round approach to remove low-quality cells. We
first use a high-resolution to get a large amount of clusters.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">obj.coembed</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing nearest neighbor graph</span></span></code></pre>
<pre><code><span><span class="co">## Computing SNN</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.coembed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">obj.coembed</span>, resolution <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">ArchR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ArchR/man/paletteDiscrete.html" class="external-link">paletteDiscrete</a></span><span class="op">(</span><span class="va">obj.coembed</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="st">"RNA_snn_res.1"</span><span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"UMAP1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"UMAP2"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-5-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can use the function <em>CellPropPlot</em> to visualize the cell
propotion across all patients.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CellPropPlot.html">CellPropPlot</a></span><span class="op">(</span><span class="va">obj.coembed</span>,</span>
<span>                  group.by <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>,</span>
<span>                  prop.in <span class="op">=</span> <span class="st">"patient_region_id"</span>,</span>
<span>                  cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-6-1.png" width="960" style="display: block; margin: auto;"></p>
<p>Since we have annotated all patients into three major groups, i.e.,
myogenic, ischmeic, and fibrotic. we can also perform statistical test
to check if any sub-population are enriched in any of the above group.
This can be done by the function <em>CompareCellProp</em>.</p>
<p>The idea is to identify sub-populations that show significant
difference.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.coembed</span><span class="op">$</span><span class="va">patient_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">obj.coembed</span><span class="op">$</span><span class="va">patient_group</span>, </span>
<span>                                    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"myogenic"</span>, <span class="st">"ischemic"</span>, <span class="st">"fibrotic"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CompareCellProp.html">CompareCellProp</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj.coembed</span>, </span>
<span>                     group.by <span class="op">=</span> <span class="st">"RNA_snn_res.1"</span>, </span>
<span>                     prop.in <span class="op">=</span> <span class="st">"patient_region_id"</span>, </span>
<span>                      sample.annotation <span class="op">=</span> <span class="st">"patient_group"</span>,</span>
<span>                    comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"myogenic"</span>, <span class="st">"ischemic"</span><span class="op">)</span>,</span>
<span>                                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ischemic"</span>, <span class="st">"fibrotic"</span><span class="op">)</span>,</span>
<span>                                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"myogenic"</span>, <span class="st">"fibrotic"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Computation failed in `stat_signif()`.</span></span>
<span><span class="co">## Caused by error in `wilcox.test.default()`:</span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> not enough 'y' observations</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Computation failed in `stat_signif()`.</span></span>
<span><span class="co">## Caused by error in `wilcox.test.default()`:</span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> not enough 'y' observations</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Computation failed in `stat_signif()`.</span></span>
<span><span class="co">## Caused by error in `wilcox.test.default()`:</span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> not enough 'y' observations</span></span></code></pre>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-7-1.png" width="1920" style="display: block; margin: auto;"></p>
<p>We subset the object to only keep the clusters with significant
difference and re-do the clustering</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">obj.coembed</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA_snn_res.1"</span></span>
<span><span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">obj.coembed</span>, idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">11</span>, <span class="fl">12</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">7</span>, <span class="fl">8</span>, <span class="fl">9</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coembed.sub</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 211895 features across 35705 samples within 3 assays </span></span>
<span><span class="co">## Active assay: RNA (19026 features, 1435 variable features)</span></span>
<span><span class="co">##  5 layers present: data, counts.2, scale.data.2, counts, scale.data</span></span>
<span><span class="co">##  2 other assays present: ATAC, GeneActivity</span></span>
<span><span class="co">##  4 dimensional reductions calculated: pca, umap, harmony, umap_harmony</span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># re-do the UMAP</span></span>
<span><span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">coembed.sub</span>, </span>
<span>                       dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, </span>
<span>                       reduction <span class="op">=</span> <span class="st">'harmony'</span>,</span>
<span>                       reduction.name <span class="op">=</span> <span class="st">"umap_harmony"</span>,</span>
<span>                       reduction.key <span class="op">=</span> <span class="st">'umap_harmony_'</span>,</span>
<span>                       verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Keys should be one or more alphanumeric characters followed by an</span></span>
<span><span class="co">## underscore, setting key from umap_harmony_ to umapharmony_</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## re-clustering with a lower resolution</span></span>
<span><span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">coembed.sub</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing nearest neighbor graph</span></span></code></pre>
<pre><code><span><span class="co">## Computing SNN</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">coembed.sub</span>, resolution <span class="op">=</span> <span class="fl">0.2</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">ArchR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ArchR/man/paletteDiscrete.html" class="external-link">paletteDiscrete</a></span><span class="op">(</span><span class="va">coembed.sub</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="st">"RNA_snn_res.0.2"</span><span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"UMAP1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"UMAP2"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">p</span>                </span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Visualize the patient and RNA/ATAC distribution</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, group.by <span class="op">=</span> <span class="st">"patient"</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, group.by <span class="op">=</span> <span class="st">"tech"</span>, reduction <span class="op">=</span> <span class="st">"umap_harmony"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-9-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>Next, we identify the markers for each cluster and visualize the top
10.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">coembed.sub</span>, </span>
<span>                              only.pos <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                              min.pct <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                              logfc.threshold <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Calculating cluster 0</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster 1</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster 2</span></span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="va">all.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, order_by <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Scaling data with a low number of groups may produce misleading</span></span>
<span><span class="co">## results</span></span></code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-10-1.png" width="1440" style="display: block; margin: auto;"></p>
<p>The above dot plot demonstrates the top 10 markers per cluster and we
can easily classify cluster 1 as myofibroblasts.</p>
<p>Let’s compare the cell proportion again:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coembed.sub</span><span class="op">$</span><span class="va">patient_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">coembed.sub</span><span class="op">$</span><span class="va">patient_group</span>, </span>
<span>                                    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"myogenic"</span>, <span class="st">"ischemic"</span>, <span class="st">"fibrotic"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CompareCellProp.html">CompareCellProp</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">coembed.sub</span>, </span>
<span>                     group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, </span>
<span>                     prop.in <span class="op">=</span> <span class="st">"patient_region_id"</span>, </span>
<span>                     sample.annotation <span class="op">=</span> <span class="st">"patient_group"</span>,</span>
<span>                     comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"myogenic"</span>, <span class="st">"ischemic"</span><span class="op">)</span>,</span>
<span>                                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ischemic"</span>, <span class="st">"fibrotic"</span><span class="op">)</span>,</span>
<span>                                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"myogenic"</span>, <span class="st">"fibrotic"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-11-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Save data</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">coembed.sub</span>, <span class="st">"./Fibroblast/coembed.sub.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="trajectory-analysis">Trajectory analysis<a class="anchor" aria-label="anchor" href="#trajectory-analysis"></a>
</h2>
<p>We next identify the trajectory for myofibroblast
differentiation.</p>
<div class="section level3">
<h3 id="dimensionality-reduction">Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h3>
<p>To infer trajectory, we will perform dimension reduction using
diffusion map via the function <em>RunDiffusionMap</em>. This is based
on the R package <a href="https://bioconductor.org/packages/release/bioc/html/destiny.html" class="external-link">destiny</a>.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coembed.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunDiffusionMap.html">RunDiffusionMap</a></span><span class="op">(</span><span class="va">coembed.sub</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## finding knns......done. Time: 395.14s</span></span>
<span><span class="co">## Calculating transition probabilities...</span></span></code></pre>
<pre><code><span><span class="co">## Warning: 'as(&lt;dsCMatrix&gt;, "dsTMatrix")' is deprecated.</span></span>
<span><span class="co">## Use 'as(., "TsparseMatrix")' instead.</span></span>
<span><span class="co">## See help("Deprecated") and help("Matrix-deprecated").</span></span></code></pre>
<pre><code><span><span class="co">## ...done. Time: 0.37s</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## performing eigen decomposition......done. Time: 3.51s</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">ArchR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ArchR/man/paletteDiscrete.html" class="external-link">paletteDiscrete</a></span><span class="op">(</span><span class="va">coembed.sub</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="st">"RNA_snn_res.0.2"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             reduction <span class="op">=</span> <span class="st">"dm"</span>, shuffle <span class="op">=</span> <span class="cn">TRUE</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"DC 1"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"DC 2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;"></p>
<p>We can also plot snATAC-seq and snRNA-seq individually</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub</span>, reduction <span class="op">=</span> <span class="st">"dm"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, split.by <span class="op">=</span> <span class="st">"tech"</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-14-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="cell-pairing">Cell pairing<a class="anchor" aria-label="anchor" href="#cell-pairing"></a>
</h3>
<p>Next, we match the cells between these two modalities. In other
words, for each cell in, for example, snATAC-seq, we will find a cell
from snRNA-seq data so that these two cells have the similar profiles.
This is only necessary when each modality was performed independently.
If snRNA-seq and snATAC-seq data was generated by multi-modal protocol,
e.g., 10X multiome or SHARE-seq, this step can be skipped.</p>
<p>We here use the method proposed by <a href="https://www.sciencedirect.com/science/article/pii/S2666979X22001082?via%3Dihub" class="external-link">Kartha,
Vinay K., et al.</a> to match the cells.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df.pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PairCells.html">PairCells</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">coembed.sub</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                     pair.mode <span class="op">=</span> <span class="st">"greedy"</span>,</span>
<span>                     pair.by <span class="op">=</span> <span class="st">"tech"</span>, ident1 <span class="op">=</span> <span class="st">"ATAC"</span>, ident2 <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Getting dimensional reduction data for pairing cells...</span></span></code></pre>
<pre><code><span><span class="co">## ATAC assay doesn't leave any cells, so it is removed</span></span></code></pre>
<pre><code><span><span class="co">## GeneActivity assay doesn't leave any cells, so it is removed</span></span></code></pre>
<pre><code><span><span class="co">## Number of ATAC cells: 4107, number of RNA cells: 31598</span></span></code></pre>
<pre><code><span><span class="co">## Pairing cells using greedy mode...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Pairing all ATAC cells to nearest RNA cells"</span></span></code></pre>
<pre><code><span><span class="co">## Finished!</span></span></code></pre>
<p>We can visualize the paired cells</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sel_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">df.pair</span><span class="op">$</span><span class="va">ATAC</span>, <span class="va">df.pair</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span></span>
<span><span class="va">coembed.sub2</span> <span class="op">&lt;-</span> <span class="va">coembed.sub</span><span class="op">[</span>, <span class="va">sel_cells</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.height <span class="op">=</span> <span class="fl">5</span>, repr.plot.width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">coembed.sub2</span>, reduction <span class="op">=</span> <span class="st">"dm"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, split.by <span class="op">=</span> <span class="st">"tech"</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-16-1.png" width="960" style="display: block; margin: auto;"></p>
<p>We next create a new Seurat object for there paired cells as if they
are generated by single-cell multimodal protocol.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreatePairedObject.html">CreatePairedObject</a></span><span class="op">(</span>df.pair <span class="op">=</span> <span class="va">df.pair</span>, </span>
<span>                               obj.coembed <span class="op">=</span> <span class="va">coembed.sub2</span>,</span>
<span>                               obj.rna <span class="op">=</span> <span class="va">obj.rna</span>,</span>
<span>                               obj.atac <span class="op">=</span> <span class="va">obj.atac</span>,</span>
<span>                               rna.assay <span class="op">=</span> <span class="st">"RNA"</span>, </span>
<span>                               atac.assay <span class="op">=</span> <span class="st">"ATAC"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Merging objects...</span></span></code></pre>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">obj.pair</span><span class="op">)</span>                           </span></code></pre></div>
<pre><code><span><span class="co">## Normalizing layer: counts</span></span></code></pre>
<p>Finally, we infer a pseudo-time trajectory from SCARA5+ fibroblasts
to myofibroblast using the approach from <a href="https://www.archrproject.com/" class="external-link">ArchR</a>. Here we modified the
function to allow to take a Seurat object as input</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj.pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTrajectory.html">AddTrajectory</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj.pair</span>, </span>
<span>                          trajectory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                          group.by <span class="op">=</span> <span class="st">"RNA_snn_res.0.2"</span>, </span>
<span>                          reduction <span class="op">=</span> <span class="st">"dm"</span>,</span>
<span>                          dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, </span>
<span>                          use.all <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>                          </span>
<span><span class="co"># we only plot the cells that are in this trajectory</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="va">obj.pair</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">obj.pair</span><span class="op">$</span><span class="va">Trajectory</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TrajectoryPlot.html">TrajectoryPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, </span>
<span>                    reduction <span class="op">=</span> <span class="st">"dm"</span>,</span>
<span>                    continuousSet <span class="op">=</span> <span class="st">"blueYellow"</span>,</span>
<span>                    size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   addArrow <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-18-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="tf-and-gene-selection">TF and gene selection<a class="anchor" aria-label="anchor" href="#tf-and-gene-selection"></a>
</h2>
<p>We next select candidate TFs and genes for building a meaningful gene
regulatory network.</p>
<div class="section level3">
<h3 id="select-tfs">Select TFs<a class="anchor" aria-label="anchor" href="#select-tfs"></a>
</h3>
<p>To identify potential regulator (i.e., TFs), we first estimate an
acitivty score for each TF in each cell. This is done by first
performing motif matching and then computing deviation scores using <a href="https://greenleaflab.github.io/chromVAR/index.html" class="external-link">chromVAR</a>.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a list of motif position frequency matrices from the JASPAR database</span></span>
<span><span class="va">pfm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TFBSTools/man/getMatrixSet-methods.html" class="external-link">getMatrixSet</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">JASPAR2020</span>,</span>
<span>  opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>collection <span class="op">=</span> <span class="st">"CORE"</span>, tax_group <span class="op">=</span> <span class="st">'vertebrates'</span>, all_versions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add motif information</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/AddMotifs.html" class="external-link">AddMotifs</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">obj</span>,</span>
<span>  genome <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span>,</span>
<span>  pfm <span class="op">=</span> <span class="va">pfm</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"ATAC"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Building motif matrix</span></span></code></pre>
<pre><code><span><span class="co">## Finding motif positions</span></span></code></pre>
<pre><code><span><span class="co">## Creating Motif object</span></span></code></pre>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunChromVAR.html" class="external-link">RunChromVAR</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">obj</span>,</span>
<span>  genome <span class="op">=</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"ATAC"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing GC bias per region</span></span></code></pre>
<pre><code><span><span class="co">## Selecting background regions</span></span></code></pre>
<pre><code><span><span class="co">## Computing deviations from background</span></span></code></pre>
<pre><code><span><span class="co">## Constructing chromVAR assay</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Layer counts isn't present in the assay object; returning NULL</span></span></code></pre>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SelectTFs.html">SelectTFs</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, return.heatmap <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Some values are below 0, this could be the Motif activity matrix in which scaleTo should be set = NULL.</span></span>
<span><span class="co">## Continuing without depth normalization!</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<pre><code><span><span class="co">## Find 456 shared features!</span></span></code></pre>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df.cor</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">tfs</span></span>
<span><span class="va">ht</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">heatmap</span></span></code></pre></div>
<p>We can visualize the TF activity dynamic along the trajectory</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">draw</span><span class="op">(</span><span class="va">ht</span><span class="op">)</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-20-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="select-genes">Select genes<a class="anchor" aria-label="anchor" href="#select-genes"></a>
</h3>
<p>We will select relevant genes by first linking genes to peaks based
on the corrleation between gene expression from the snRNA-seq data and
peak accessibility from the snATAC-seq data along the inferred
trajectory. This means that we only consider a gene to be a potential
target if it can be assocaited to at least one peak.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SelectGenes.html">SelectGenes</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>,</span>
<span>                  labelTop1 <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                  labelTop2 <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<pre><code><span><span class="co">## Linking cis-regulatory elements to genes...</span></span></code></pre>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df.p2g</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">p2g</span></span>
<span><span class="va">ht</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">heatmap</span></span>
<span></span>
<span><span class="fu">draw</span><span class="op">(</span><span class="va">ht</span><span class="op">)</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-21-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="gene-regulatory-network-inference">Gene regulatory network inference<a class="anchor" aria-label="anchor" href="#gene-regulatory-network-inference"></a>
</h2>
<p>We here will try to predict a gene regulatory network based on the
correlation of the TF binding activity as estimated from snATAC-seq and
gene expression as measured by snRNA-seq along the trajectory.</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tf.gene.cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetTFGeneCorrelation.html">GetTFGeneCorrelation</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, </span>
<span>                                    tf.use <span class="op">=</span> <span class="va">df.cor</span><span class="op">$</span><span class="va">tfs</span>, </span>
<span>                                    gene.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df.p2g</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>,</span>
<span>                                    tf.assay <span class="op">=</span> <span class="st">"chromvar"</span>, </span>
<span>                                    gene.assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>                                    trajectory.name <span class="op">=</span> <span class="st">"Trajectory"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Some values are below 0, this could be the Motif activity matrix in which scaleTo should be set = NULL.</span></span>
<span><span class="co">## Continuing without depth normalization!</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<p>We can then visualize this correlation matrix by heatmap. Also, we
can cluster the genes and TFs to identify different regulatory modules
for the predefined sub-populations.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ht</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRNHeatmap.html">GRNHeatmap</a></span><span class="op">(</span><span class="va">tf.gene.cor</span>, </span>
<span>                 tf.timepoint <span class="op">=</span> <span class="va">df.cor</span><span class="op">$</span><span class="va">time_point</span><span class="op">)</span></span>
<span><span class="va">ht</span></span></code></pre></div>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-23-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>To associate genes to TFs, we will use the peak-to-gene links and TF
binding sites information. Specifically, if a gene is regulated by a
peak and this peak is bound by a TF, then we consider this gene to be a
target of this TF.</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">motif.matching</span> <span class="op">&lt;-</span> <span class="va">obj</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ATAC</span><span class="op">@</span><span class="va">motifs</span><span class="op">@</span><span class="va">data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">motif.matching</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">obj</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ATAC</span><span class="op">@</span><span class="va">motifs</span><span class="op">@</span><span class="va">motif.names</span></span>
<span><span class="va">motif.matching</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">motif.matching</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">df.p2g</span><span class="op">$</span><span class="va">peak</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">tf.gene.cor</span><span class="op">$</span><span class="va">tf</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="va">df.grn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetGRN.html">GetGRN</a></span><span class="op">(</span>motif.matching <span class="op">=</span> <span class="va">motif.matching</span>, </span>
<span>                 df.cor <span class="op">=</span> <span class="va">tf.gene.cor</span>, </span>
<span>                 df.p2g <span class="op">=</span> <span class="va">df.p2g</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Filtering network by peak-to-gene links...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering network by TF binding site prediction...</span></span></code></pre>
<p>Next, we can visualize our network as the last step of this
analysis</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define colors for nodes representing TFs (i.e., regulators)</span></span>
<span><span class="va">df.cor</span> <span class="op">&lt;-</span> <span class="va">df.cor</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">df.cor</span><span class="op">$</span><span class="va">time_point</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">tfs.timepoint</span> <span class="op">&lt;-</span> <span class="va">df.cor</span><span class="op">$</span><span class="va">time_point</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tfs.timepoint</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">df.cor</span><span class="op">$</span><span class="va">tfs</span></span>
<span></span>
<span><span class="co"># plot the graph, here we can highlight some genes</span></span>
<span><span class="va">df.grn2</span> <span class="op">&lt;-</span> <span class="va">df.grn</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">correlation</span> <span class="op">&gt;</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">tf</span>, <span class="va">gene</span>, <span class="va">correlation</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>weights <span class="op">=</span> <span class="va">correlation</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">df.grn2</span>,<span class="st">"./Fibroblast/final_grn.Rds"</span><span class="op">)</span>    </span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRNPlot.html">GRNPlot</a></span><span class="op">(</span><span class="va">df.grn2</span>, </span>
<span>             tfs.timepoint <span class="op">=</span> <span class="va">tfs.timepoint</span>,</span>
<span>             show.tf.labels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             seed <span class="op">=</span> <span class="fl">42</span>, </span>
<span>             plot.importance <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            min.importance <span class="op">=</span> <span class="fl">2</span>,</span>
<span>            remove.isolated <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repr.plot.height <span class="op">=</span> <span class="fl">20</span>, repr.plot.width <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Using alpha for a discrete variable is not advised.</span></span></code></pre>
<p><img src="myofibroblast-GRN_files/figure-html/viz_network-1.png" width="1440" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="grn-visualization">GRN visualization<a class="anchor" aria-label="anchor" href="#grn-visualization"></a>
</h2>
<div class="section level3">
<h3 id="grn-along-the-trajectory">GRN along the trajectory<a class="anchor" aria-label="anchor" href="#grn-along-the-trajectory"></a>
</h3>
<p>Once we generated the gene regulatory network, we can visualize
individual TFs in terms of binding activity, expression, and target
expression along the pseudotime trajectory.</p>
<p>Here we select two TFs for visualization.</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTargetAssay.html">AddTargetAssay</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, df.grn <span class="op">=</span> <span class="va">df.grn2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Layer counts isn't present in the assay object; returning NULL</span></span></code></pre>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PseudotimePlot.html">PseudotimePlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, tf.use <span class="op">=</span> <span class="st">"NR3C2"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Some values are below 0, this could be the Motif activity matrix in which scaleTo should be set = NULL.</span></span>
<span><span class="co">## Continuing without depth normalization!</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Some values are below 0, this could be the Motif activity matrix in which scaleTo should be set = NULL.</span></span>
<span><span class="co">## Continuing without depth normalization!</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PseudotimePlot.html">PseudotimePlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, tf.use <span class="op">=</span> <span class="st">"RUNX1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Some values are below 0, this could be the Motif activity matrix in which scaleTo should be set = NULL.</span></span>
<span><span class="co">## Continuing without depth normalization!</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<pre><code><span><span class="co">## Creating Trajectory Group Matrix..</span></span></code></pre>
<pre><code><span><span class="co">## Some values are below 0, this could be the Motif activity matrix in which scaleTo should be set = NULL.</span></span>
<span><span class="co">## Continuing without depth normalization!</span></span></code></pre>
<pre><code><span><span class="co">## Smoothing...</span></span></code></pre>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<pre><code><span><span class="co">## `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<pre><code><span><span class="co">## `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<p><img src="myofibroblast-GRN_files/figure-html/unnamed-chunk-25-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>The x-axis in above plots present pseudotime point along the
trajectory, and the y-axis represent TF binding acitivty, TF expression,
and TF target expression after z-score transformation.</p>
<p>Save data</p>
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">obj</span>, <span class="st">"./Fibroblast/coembed.trajectory.rds"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check session information</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">## Platform: x86_64-conda-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: CentOS Linux 7 (Core)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /data/pinello/SHARED_SOFTWARE/anaconda_latest/envs/zl_envs/zl_scmega/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    grid      stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] circlize_0.4.16                   ComplexHeatmap_2.18.0            </span></span>
<span><span class="co">##  [3] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.70.2                  </span></span>
<span><span class="co">##  [5] rtracklayer_1.62.0                BiocIO_1.12.0                    </span></span>
<span><span class="co">##  [7] Biostrings_2.70.3                 XVector_0.42.0                   </span></span>
<span><span class="co">##  [9] ggraph_2.2.1                      igraph_2.0.3                     </span></span>
<span><span class="co">## [11] TFBSTools_1.40.0                  JASPAR2020_0.99.10               </span></span>
<span><span class="co">## [13] dplyr_1.1.4                       Nebulosa_1.12.1                  </span></span>
<span><span class="co">## [15] patchwork_1.2.0                   harmony_1.2.0                    </span></span>
<span><span class="co">## [17] scMEGA_1.0.2                      Signac_1.13.0                    </span></span>
<span><span class="co">## [19] Seurat_5.1.0                      SeuratObject_5.0.2               </span></span>
<span><span class="co">## [21] sp_2.1-4                          rhdf5_2.46.1                     </span></span>
<span><span class="co">## [23] SummarizedExperiment_1.32.0       Biobase_2.62.0                   </span></span>
<span><span class="co">## [25] MatrixGenerics_1.14.0             Rcpp_1.0.12                      </span></span>
<span><span class="co">## [27] Matrix_1.6-5                      GenomicRanges_1.54.1             </span></span>
<span><span class="co">## [29] GenomeInfoDb_1.38.8               IRanges_2.36.0                   </span></span>
<span><span class="co">## [31] S4Vectors_0.40.2                  BiocGenerics_0.48.1              </span></span>
<span><span class="co">## [33] matrixStats_1.3.0                 data.table_1.15.4                </span></span>
<span><span class="co">## [35] stringr_1.5.1                     plyr_1.8.9                       </span></span>
<span><span class="co">## [37] magrittr_2.0.3                    ggplot2_3.5.1                    </span></span>
<span><span class="co">## [39] gtable_0.3.5                      gtools_3.9.5                     </span></span>
<span><span class="co">## [41] gridExtra_2.3                     ArchR_1.0.2                      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] R.methodsS3_1.8.2           nnet_7.3-19                </span></span>
<span><span class="co">##   [3] poweRlaw_0.80.0             DT_0.33                    </span></span>
<span><span class="co">##   [5] goftest_1.2-3               vctrs_0.6.5                </span></span>
<span><span class="co">##   [7] spatstat.random_3.2-3       shape_1.4.6.1              </span></span>
<span><span class="co">##   [9] digest_0.6.36               png_0.1-8                  </span></span>
<span><span class="co">##  [11] proxy_0.4-27                ggrepel_0.9.5              </span></span>
<span><span class="co">##  [13] deldir_2.0-4                parallelly_1.37.1          </span></span>
<span><span class="co">##  [15] MASS_7.3-60                 pkgdown_2.0.9              </span></span>
<span><span class="co">##  [17] reshape2_1.4.4              foreach_1.5.2              </span></span>
<span><span class="co">##  [19] httpuv_1.6.15               withr_3.0.0                </span></span>
<span><span class="co">##  [21] ggrastr_1.0.2               xfun_0.45                  </span></span>
<span><span class="co">##  [23] ggpubr_0.6.0                survival_3.7-0             </span></span>
<span><span class="co">##  [25] memoise_2.0.1               hexbin_1.28.3              </span></span>
<span><span class="co">##  [27] ggbeeswarm_0.7.2            systemfonts_1.0.5          </span></span>
<span><span class="co">##  [29] GlobalOptions_0.1.2         ragg_1.3.2                 </span></span>
<span><span class="co">##  [31] zoo_1.8-12                  pbapply_1.7-2              </span></span>
<span><span class="co">##  [33] R.oo_1.26.0                 DEoptimR_1.1-3             </span></span>
<span><span class="co">##  [35] KEGGREST_1.42.0             promises_1.3.0             </span></span>
<span><span class="co">##  [37] scatterplot3d_0.3-44        httr_1.4.7                 </span></span>
<span><span class="co">##  [39] rstatix_0.7.2               restfulr_0.0.15            </span></span>
<span><span class="co">##  [41] globals_0.16.3              fitdistrplus_1.1-11        </span></span>
<span><span class="co">##  [43] rhdf5filters_1.14.1         miniUI_0.1.1.1             </span></span>
<span><span class="co">##  [45] generics_0.1.3              curl_5.2.1                 </span></span>
<span><span class="co">##  [47] zlibbioc_1.48.2             polyclip_1.10-6            </span></span>
<span><span class="co">##  [49] chromVAR_1.24.0             GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">##  [51] SparseArray_1.2.4           RcppEigen_0.3.4.0.0        </span></span>
<span><span class="co">##  [53] doParallel_1.0.17           xtable_1.8-4               </span></span>
<span><span class="co">##  [55] desc_1.4.3                  pracma_2.4.4               </span></span>
<span><span class="co">##  [57] evaluate_0.24.0             S4Arrays_1.2.1             </span></span>
<span><span class="co">##  [59] hms_1.1.3                   irlba_2.3.5.1              </span></span>
<span><span class="co">##  [61] colorspace_2.1-0            ROCR_1.0-11                </span></span>
<span><span class="co">##  [63] reticulate_1.38.0           spatstat.data_3.1-2        </span></span>
<span><span class="co">##  [65] lmtest_0.9-40               readr_2.1.5                </span></span>
<span><span class="co">##  [67] later_1.3.2                 viridis_0.6.5              </span></span>
<span><span class="co">##  [69] lattice_0.22-6              spatstat.geom_3.2-9        </span></span>
<span><span class="co">##  [71] future.apply_1.11.2         robustbase_0.99-2          </span></span>
<span><span class="co">##  [73] scattermore_1.2             XML_3.99-0.17              </span></span>
<span><span class="co">##  [75] cowplot_1.1.3               RcppAnnoy_0.0.22           </span></span>
<span><span class="co">##  [77] xts_0.14.0                  class_7.3-22               </span></span>
<span><span class="co">##  [79] pillar_1.9.0                nlme_3.1-165               </span></span>
<span><span class="co">##  [81] iterators_1.0.14            caTools_1.18.2             </span></span>
<span><span class="co">##  [83] compiler_4.3.3              RSpectra_0.16-1            </span></span>
<span><span class="co">##  [85] stringi_1.8.4               tensor_1.5                 </span></span>
<span><span class="co">##  [87] GenomicAlignments_1.38.2    crayon_1.5.3               </span></span>
<span><span class="co">##  [89] abind_1.4-5                 graphlayouts_1.1.1         </span></span>
<span><span class="co">##  [91] bit_4.0.5                   pcaMethods_1.94.0          </span></span>
<span><span class="co">##  [93] fastmatch_1.1-4             codetools_0.2-20           </span></span>
<span><span class="co">##  [95] textshaping_0.3.7           TTR_0.24.4                 </span></span>
<span><span class="co">##  [97] bslib_0.7.0                 textshape_1.7.5            </span></span>
<span><span class="co">##  [99] e1071_1.7-14                destiny_3.16.0             </span></span>
<span><span class="co">## [101] GetoptLong_1.0.5            ggplot.multistats_1.0.0    </span></span>
<span><span class="co">## [103] plotly_4.10.4               mime_0.12                  </span></span>
<span><span class="co">## [105] splines_4.3.3               fastDummies_1.7.3          </span></span>
<span><span class="co">## [107] knitr_1.47                  blob_1.2.4                 </span></span>
<span><span class="co">## [109] utf8_1.2.4                  clue_0.3-65                </span></span>
<span><span class="co">## [111] seqLogo_1.68.0              fs_1.6.4                   </span></span>
<span><span class="co">## [113] listenv_0.9.1               ggsignif_0.6.4             </span></span>
<span><span class="co">## [115] tibble_3.2.1                tzdb_0.4.0                 </span></span>
<span><span class="co">## [117] tweenr_2.0.3                pkgconfig_2.0.3            </span></span>
<span><span class="co">## [119] tools_4.3.3                 cachem_1.1.0               </span></span>
<span><span class="co">## [121] RhpcBLASctl_0.23-42         RSQLite_2.3.7              </span></span>
<span><span class="co">## [123] viridisLite_0.4.2           smoother_1.3               </span></span>
<span><span class="co">## [125] DBI_1.2.3                   fastmap_1.2.0              </span></span>
<span><span class="co">## [127] rmarkdown_2.27              scales_1.3.0               </span></span>
<span><span class="co">## [129] ica_1.0-3                   Rsamtools_2.18.0           </span></span>
<span><span class="co">## [131] broom_1.0.6                 sass_0.4.9                 </span></span>
<span><span class="co">## [133] FNN_1.1.4                   dotCall64_1.1-1            </span></span>
<span><span class="co">## [135] carData_3.0-5               RANN_2.6.1                 </span></span>
<span><span class="co">## [137] farver_2.1.2                mgcv_1.9-1                 </span></span>
<span><span class="co">## [139] tidygraph_1.3.1             yaml_2.3.8                 </span></span>
<span><span class="co">## [141] ggthemes_5.1.0              cli_3.6.3                  </span></span>
<span><span class="co">## [143] purrr_1.0.2                 motifmatchr_1.24.0         </span></span>
<span><span class="co">## [145] leiden_0.4.3.1              lifecycle_1.0.4            </span></span>
<span><span class="co">## [147] uwot_0.2.2                  mvtnorm_1.2-5              </span></span>
<span><span class="co">## [149] presto_1.0.0                backports_1.5.0            </span></span>
<span><span class="co">## [151] BiocParallel_1.36.0         annotate_1.80.0            </span></span>
<span><span class="co">## [153] rjson_0.2.21                ggridges_0.5.6             </span></span>
<span><span class="co">## [155] progressr_0.14.0            parallel_4.3.3             </span></span>
<span><span class="co">## [157] jsonlite_1.8.8              RcppHNSW_0.6.0             </span></span>
<span><span class="co">## [159] bitops_1.0-7                bit64_4.0.5                </span></span>
<span><span class="co">## [161] Rtsne_0.17                  spatstat.utils_3.0-5       </span></span>
<span><span class="co">## [163] nabor_0.5.0                 ranger_0.16.0              </span></span>
<span><span class="co">## [165] CNEr_1.38.0                 jquerylib_0.1.4            </span></span>
<span><span class="co">## [167] highr_0.11                  knn.covertree_1.0          </span></span>
<span><span class="co">## [169] R.utils_2.12.3              lazyeval_0.2.2             </span></span>
<span><span class="co">## [171] shiny_1.8.1.1               htmltools_0.5.8.1          </span></span>
<span><span class="co">## [173] GO.db_3.18.0                sctransform_0.4.1          </span></span>
<span><span class="co">## [175] glue_1.7.0                  factoextra_1.0.7           </span></span>
<span><span class="co">## [177] TFMPvalue_0.0.9             spam_2.10-0                </span></span>
<span><span class="co">## [179] VIM_6.2.2                   RCurl_1.98-1.14            </span></span>
<span><span class="co">## [181] mclust_6.1.1                ks_1.14.2                  </span></span>
<span><span class="co">## [183] boot_1.3-30                 R6_2.5.1                   </span></span>
<span><span class="co">## [185] tidyr_1.3.1                 SingleCellExperiment_1.24.0</span></span>
<span><span class="co">## [187] RcppRoll_0.3.0              vcd_1.4-12                 </span></span>
<span><span class="co">## [189] labeling_0.4.3              cluster_2.1.6              </span></span>
<span><span class="co">## [191] Rhdf5lib_1.24.2             DirichletMultinomial_1.44.0</span></span>
<span><span class="co">## [193] DelayedArray_0.28.0         tidyselect_1.2.1           </span></span>
<span><span class="co">## [195] vipor_0.4.7                 ggforce_0.4.2              </span></span>
<span><span class="co">## [197] car_3.1-2                   AnnotationDbi_1.64.1       </span></span>
<span><span class="co">## [199] future_1.33.2               munsell_0.5.1              </span></span>
<span><span class="co">## [201] KernSmooth_2.23-24          laeken_0.5.3               </span></span>
<span><span class="co">## [203] htmlwidgets_1.6.4           RColorBrewer_1.1-3         </span></span>
<span><span class="co">## [205] rlang_1.1.4                 spatstat.sparse_3.1-0      </span></span>
<span><span class="co">## [207] spatstat.explore_3.2-7      fansi_1.0.6                </span></span>
<span><span class="co">## [209] Cairo_1.6-2                 beeswarm_0.4.0</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zhijian Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
