<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene-regulatory network of CD4 T cells activaction • scMEGA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene-regulatory network of CD4 T cells activaction">
<meta property="og:description" content="scMEGA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMEGA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/myofibroblast-GRN.html">Infer GRN for myofibroblast</a>
    </li>
    <li>
      <a href="../articles/cardiomyocyte-GRN.html">Infer GRN for cardiomyocytes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/CostaLab/scMEGA" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene-regulatory network of CD4 T cells activaction</h1>
            
            <h4 data-toc-skip class="date">Compiled: November 30, 2022</h4>
      
      
      <div class="hidden name"><code>pbmc_10x_multiome.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we’ll demonstrate how to infer gene regulatory network using single-cell multimodal data which measures gene expression and chromatin accessibility profiles from the same single cells. We will use a publicly available 10x Genomic Multiome data set for <a href="https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-10-k-1-standard-2-0-0" class="external-link">human PBMCs</a>.</p>
<p>We first download the required data.</p>
<p>Run the following commands to download the data:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> 10x_pbmc</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-q</span> <span class="at">-P</span> 10x_pbmc https://cf.10xgenomics.com/samples/cell-arc/2.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-q</span> <span class="at">-P</span> 10x_pbmc https://cf.10xgenomics.com/samples/cell-arc/2.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-q</span> <span class="at">-P</span> 10x_pbmc https://cf.10xgenomics.com/samples/cell-arc/2.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5</span></code></pre></div>
<p>Next, we load all necessary packages:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ArchR</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://costalab.github.io/scMEGA/">scMEGA</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/powellgenomicslab/Nebulosa" class="external-link">Nebulosa</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomeInfoDb" class="external-link">GenomeInfoDb</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://jaspar.genereg.net/" class="external-link">JASPAR2020</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ge11232002/TFBSTools" class="external-link">TFBSTools</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org" class="external-link">igraph</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="data-pre-processing">Data pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"></a>
</h2>
<p>Let’s load the data into memory and extract scRNA and scATAC-seq data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">inputdata.10x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span><span class="st">"./10x_pbmc/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Genome matrix has multiple modalities, returning a list of matrices for this genome</span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extract RNA and ATAC data</span>
<span class="va">rna_counts</span> <span class="op">&lt;-</span> <span class="va">inputdata.10x</span><span class="op">$</span><span class="va">`Gene Expression`</span>
<span class="va">atac_counts</span> <span class="op">&lt;-</span> <span class="va">inputdata.10x</span><span class="op">$</span><span class="va">Peaks</span>

<span class="co"># filter peaks by chromosome</span>
<span class="va">atac_counts</span> <span class="op">&lt;-</span> <span class="va">atac_counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">atac_counts</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></code></pre></div>
<p>Next create a Seurat object and filter low-quality cells in scRNA-seq data</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">rna_counts</span><span class="op">)</span>
<span class="va">obj.rna</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">obj.rna</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></code></pre></div>
<p>We can visualize the data quality</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Visualize QC metrics as a violin plot</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">obj.rna</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span>, pt.size <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></code></pre></div>
<p><img src="pbmc_10x_multiome_files/figure-html/unnamed-chunk-5-1.png" width="1440" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">obj.rna</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">3000</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">20</span><span class="op">)</span>

<span class="va">obj.rna</span></code></pre></div>
<pre><code><span class="co">## An object of class Seurat </span>
<span class="co">## 36601 features across 10504 samples within 1 assay </span>
<span class="co">## Active assay: RNA (36601 features, 0 variable features)</span></code></pre>
<p>Create Seurat object for scATAC-seq data:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create seurat object</span>
<span class="va">chrom_assay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/signac/reference/CreateChromatinAssay.html" class="external-link">CreateChromatinAssay</a></span><span class="op">(</span>
    counts <span class="op">=</span> <span class="va">atac_counts</span>,
    sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">":"</span>, <span class="st">"-"</span><span class="op">)</span>,
    min.cells <span class="op">=</span> <span class="fl">1</span>,
    genome <span class="op">=</span> <span class="st">'hg38'</span>,
    fragments <span class="op">=</span> <span class="st">'./10x_pbmc/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz'</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Computing hash</span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>
    counts <span class="op">=</span> <span class="va">chrom_assay</span>,
    assay <span class="op">=</span> <span class="st">"ATAC"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: Keys should be one or more alphanumeric characters followed by an</span>
<span class="co">## underscore, setting key from atac to atac_</span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extract gene annotations from EnsDb</span>
<span class="va">annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/signac/reference/GetGRangesFromEnsDb.html" class="external-link">GetGRangesFromEnsDb</a></span><span class="op">(</span>ensdb <span class="op">=</span> <span class="va">EnsDb.Hsapiens.v86</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span></code></pre>
<pre><code><span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span>

<span class="co">## Warning in .Seqinfo.mergexy(x, y): The 2 combined objects have no sequence levels in common. (Use</span>
<span class="co">##   suppressWarnings() to suppress this warning.)</span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># change to UCSC style since the data was mapped to hg38</span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">annotations</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'UCSC'</span>

<span class="co"># add the gene information to the object</span>
<span class="fu"><a href="https://satijalab.org/signac/reference/Annotation.html" class="external-link">Annotation</a></span><span class="op">(</span><span class="va">obj.atac</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">annotations</span></code></pre></div>
<p>Make sure that the two modalities have the same cells:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cell.sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">obj.rna</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">obj.atac</span><span class="op">)</span><span class="op">)</span>

<span class="va">obj.rna</span> <span class="op">&lt;-</span> <span class="va">obj.rna</span><span class="op">[</span>, <span class="va">cell.sel</span><span class="op">]</span>
<span class="va">obj.atac</span> <span class="op">&lt;-</span> <span class="va">obj.atac</span><span class="op">[</span>, <span class="va">cell.sel</span><span class="op">]</span></code></pre></div>
<p>We next process the data using standard Seurat analysis pipeline:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.rna</span> <span class="op">&lt;-</span> <span class="va">obj.rna</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
<span class="co">## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span>
<span class="co">## This message will be shown once per session</span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj.atac</span> <span class="op">&lt;-</span> <span class="va">obj.atac</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://satijalab.org/signac/reference/RunTFIDF.html" class="external-link">RunTFIDF</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://satijalab.org/signac/reference/FindTopFeatures.html" class="external-link">FindTopFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://satijalab.org/signac/reference/RunSVD.html" class="external-link">RunSVD</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
    <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">'lsi'</span>, dims <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Performing TF-IDF normalization</span></code></pre>
<pre><code><span class="co">## Running SVD</span></code></pre>
<pre><code><span class="co">## Scaling cell embeddings</span></code></pre>
<p>Visualize the scRNA-seq and scATAC-seq separately.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.rna</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj.atac</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></code></pre></div>
<p><img src="pbmc_10x_multiome_files/figure-html/unnamed-chunk-9-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zhijian Li.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
